## ALT Corpus Translation Tutorial

### NMT Framework: [https://github.com/marian-nmt/marian](https://github.com/marian-nmt/marian)  
### ALT Corpus Information: [https://www2.nict.go.jp/astrec-att/member/mutiyama/ALT/](https://www2.nict.go.jp/astrec-att/member/mutiyama/ALT/)  

Date: 26 Sept 2026 and 27 Sept 2026  
By Ye Kyaw Thu, Lab Leader, Language Understaning Lab, AI Mario, Myanmar  


## Data Preparation

In [3]:
%cd /mnt/disk1/ye/exp/

/mnt/disk1/ye/exp


In [4]:
!mkdir alt-tran

In [5]:
%cd alt-tran

/mnt/disk1/ye/exp/alt-tran


In [6]:
!wget https://www2.nict.go.jp/astrec-att/member/mutiyama/ALT/ALT-Parallel-Corpus-20191206.zip

--2026-09-26 11:59:31--  https://www2.nict.go.jp/astrec-att/member/mutiyama/ALT/ALT-Parallel-Corpus-20191206.zip
Resolving www2.nict.go.jp (www2.nict.go.jp)... 133.243.3.26, 2001:df0:232:300::1a
Connecting to www2.nict.go.jp (www2.nict.go.jp)|133.243.3.26|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 21105607 (20M) [application/zip]
Saving to: ‘ALT-Parallel-Corpus-20191206.zip’

ALT-Parallel-Corpus 100%[===================>]  20.13M  1.20MB/s    in 18s     

2026-09-26 11:59:49 (1.15 MB/s) - ‘ALT-Parallel-Corpus-20191206.zip’ saved [21105607/21105607]



In [7]:
!ls

ALT-Parallel-Corpus-20191206.zip


In [8]:
!unzip ALT-Parallel-Corpus-20191206.zip

Archive:  ALT-Parallel-Corpus-20191206.zip
   creating: ALT-Parallel-Corpus-20191206/
  inflating: ALT-Parallel-Corpus-20191206/ALT-O-COCOSDA.pdf  
  inflating: ALT-Parallel-Corpus-20191206/URL.txt  
  inflating: ALT-Parallel-Corpus-20191206/data_en.txt  
  inflating: ALT-Parallel-Corpus-20191206/data_en_tok.txt  
  inflating: ALT-Parallel-Corpus-20191206/data_fil.txt  
  inflating: ALT-Parallel-Corpus-20191206/data_id.txt  
  inflating: ALT-Parallel-Corpus-20191206/data_khm.txt  
  inflating: ALT-Parallel-Corpus-20191206/data_lo.txt  
  inflating: ALT-Parallel-Corpus-20191206/data_ms.txt  
  inflating: ALT-Parallel-Corpus-20191206/data_my.txt  
  inflating: ALT-Parallel-Corpus-20191206/data_th.txt  
  inflating: ALT-Parallel-Corpus-20191206/data_vi.txt  
  inflating: ALT-Parallel-Corpus-20191206/data_bg.txt  
  inflating: ALT-Parallel-Corpus-20191206/data_zh.txt  
  inflating: ALT-Parallel-Corpus-20191206/data_ja.txt  
  inflating: ALT-Parallel-Corpus-20191206/ChangeLog.txt  
  inflat

In [12]:
%cd ALT-Parallel-Corpus-20191206/

/mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206


In [14]:
!head URL.txt

URL.100161	http://en.wikinews.org/wiki/Steve_Wright,_killer_of_five_women_in_Suffolk,_England,_sentenced_to_life_imprisonment
URL.100196	http://en.wikinews.org/wiki/Wikinews_interviews_U.S._Libertarian_presidential_candidate_James_Burns
URL.10022	http://en.wikinews.org/wiki/Kahne_takes_Checkered_Flag_at_Richmond
URL.1002218	http://en.wikinews.org/wiki/Pickens_County,_South_Carolina_sheriff_refuses_to_lower_US_flag_to_honor_Mandela
URL.100309	http://en.wikinews.org/wiki/Ralph_Nader_enters_US_presidential_race_as_independent
URL.100473	http://en.wikinews.org/wiki/Microsoft_Network_users_experience_international_outage
URL.100491	http://en.wikinews.org/wiki/Iraq_demands_immediate_withdrawal_of_Turkish_troops
URL.100527	http://en.wikinews.org/wiki/EU_fines_Microsoft_$1.35_billion_for_non-compliance_with_antitrust_decision
URL.1005448	http://en.wikinews.org/wiki/Comic_Relief_funds_allegedly_invested_in_arms,_alcohol_and_tobacco_firms
URL.10077	http://en.wikinews.org/wiki/Microsoft_develops_

In [15]:
!head data*.txt

==> data_bg.txt <==
SNT.80188.1	ফ্রান্সের প্যারিসের পার্ক দি প্রিন্সেস-এ হওয়া ২০০৭-এর রাগবি বিশ্বকাপের পুল সি-তে ইটালি পর্তুগালকে ৩১-৫ গোলে হারিয়েছে।
SNT.80188.2	অ্যান্ড্রিয়া মেসি চার মিনিটের মাথায় ট্রাইব্রেকারের মাধ্যমে ইটালির পক্ষে একটি গোল স্কোর করেন।
SNT.80188.3	প্রথমার্ধের বেশির ভাগ সময় ধরেই খেলা নিজেদের নিয়ন্ত্রণে রাখা সত্ত্বেও ইটালি বিরতির আগে আর কোনো ট্রাই স্কোর করতে পারেনি, কিন্তু ডেভিড বোর্তোলুচ্চি তিনটি পেনাল্টি মেরে তাদের এগিয়ে থাকাকে ধরে রেখেছিলেন।
SNT.80188.4	পর্তুগাল কখনোই হাল ছেড়ে দেয়নি ও ৩৩ মিনিটের মাথায় ডেভিড পেনালভা একটি ট্রাই স্কোর করেছেন, যা কিনা খেলায় তাদের একমাত্র পয়েন্ট এনে দিয়েছে।
SNT.80188.5	প্রথমার্ধে ইটালি ১৬-৫-এ এগিয়ে ছিল কিন্তু দ্বিতীয়ার্ধে বেশির ভাগ সময়েই পর্তুগাল তাদের সমানে টক্কর দিয়েছে।
SNT.80188.6	ততসত্ত্বেও বোর্তোলুচ্চি তার ম্যাচের চতুর্থ পেনাল্টিটি মারেন, এবং তারপরে মাউরো বের্গামাস্কোর ও অ্যান্ড্রিয়া মেসির দ্বিতীয় ট্রাই ইটালির জিত নিশ্চিত করে।
SNT.80188.7	বর্তমানে ৮ পয়েন্ট পেয়ে পুল সি-তে তৃতীয় স্থানে থাকা ইটালিকে ২৯ সেপ্টেম্বর দ্বিতীয় স্থানে থাকা স্

In [16]:
!wc data*.txt

   20106   433526  7671448 data_bg.txt
   20106   542339  3212272 data_en_tok.txt
   20106   478337  3129072 data_en.txt
   20106   561283  3672459 data_fil.txt
   20106   550475  7386461 data_hi.txt
   20106   441390  3395565 data_id.txt
   20106    40207  3858896 data_ja.txt
   20106   119190  8757251 data_khm.txt
   20106   141763  7735384 data_lo.txt
   20106   442878  3435020 data_ms.txt
   20106   550581 10917786 data_my.txt
   20106   143131  7528232 data_th.txt
   20106   661379  4196346 data_vi.txt
   20106    40608  2825676 data_zh.txt
  281484  5147087 77721868 total


## Preprocessing Codes

In [17]:
!ls *.py

split_corpus.py


```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
split_corpus.py  --  Split one corpus file into train / valid / test.

Usage:
    python3 split_corpus.py <input_file> <output_dir> <lang_code> [options]

Example:
    python3 split_corpus.py data_ja.txt ./split ja
        ->  ./split/train.ja   ./split/valid.ja   ./split/test.ja

Notes:
  * Uses a FIXED random seed (default 42) so that, given the same number
    of input lines, every language gets the IDENTICAL split.  This keeps
    the parallel corpus aligned.
  * Ratios default to 70% / 20% / 10%.  Test gets the remainder so that
    train+valid+test == total, always.
"""

import argparse
import os
import random


DEFAULT_TRAIN = 0.70
DEFAULT_VALID = 0.20
DEFAULT_SEED  = 42


def split_indices(n: int, train_ratio: float, valid_ratio: float, seed: int):
    """Return (train_idx, valid_idx, test_idx) as sorted index lists."""
    idx = list(range(n))
    random.seed(seed)                    # deterministic shuffle
    random.shuffle(idx)

    n_train = int(n * train_ratio)
    n_valid = int(n * valid_ratio)
    # test takes the remainder -> guarantees the three sets sum to n
    train = sorted(idx[:n_train])
    valid = sorted(idx[n_train:n_train + n_valid])
    test  = sorted(idx[n_train + n_valid:])
    return train, valid, test


def write_subset(lines, indices, path: str) -> None:
    with open(path, "w", encoding="utf-8") as f:
        for i in indices:
            f.write(lines[i])


def main() -> None:
    ap = argparse.ArgumentParser(description="Split one corpus file into train/valid/test.")
    ap.add_argument("input",      help="input corpus file, e.g. data_ja.txt")
    ap.add_argument("out_dir",    help="output directory, e.g. ./split")
    ap.add_argument("lang",       help="language code for the extension, e.g. ja, my, kh")
    ap.add_argument("--train", type=float, default=DEFAULT_TRAIN, help="train ratio (default 0.70)")
    ap.add_argument("--valid", type=float, default=DEFAULT_VALID, help="valid ratio (default 0.20)")
    ap.add_argument("--seed",  type=int,   default=DEFAULT_SEED,  help="random seed (default 42)")
    ap.add_argument("--no-shuffle", action="store_true",
                    help="disable shuffling (use contiguous slices instead)")
    args = ap.parse_args()

    os.makedirs(args.out_dir, exist_ok=True)

    with open(args.input, "r", encoding="utf-8") as f:
        lines = f.readlines()
    n = len(lines)

    if args.no_shuffle:
        n_train = int(n * args.train)
        n_valid = int(n * args.valid)
        train_i = list(range(0, n_train))
        valid_i = list(range(n_train, n_train + n_valid))
        test_i  = list(range(n_train + n_valid, n))
    else:
        train_i, valid_i, test_i = split_indices(n, args.train, args.valid, args.seed)

    base = args.out_dir.rstrip("/")
    train_path = f"{base}/train.{args.lang}"
    valid_path = f"{base}/valid.{args.lang}"
    test_path  = f"{base}/test.{args.lang}"

    write_subset(lines, train_i, train_path)
    write_subset(lines, valid_i, valid_path)
    write_subset(lines, test_i,  test_path)

    print(f"[{args.lang}] total={n:>6}  train={len(train_i):>6}  "
          f"valid={len(valid_i):>6}  test={len(test_i):>6}  "
          f"-> {train_path}, {valid_path}, {test_path}")


if __name__ == "__main__":
    main()

```

In [20]:
!ls *.sh

split_all.sh  strip_first_column.sh


```bash
#!/bin/bash
# split_all.sh -- split every language of the ALT parallel corpus.
set -euo pipefail

INDIR=/mnt/disk1/ye/data/ALT-Parallel-Corpus-20191206
OUTDIR=./split
SEED=42

mkdir -p "$OUTDIR"

# basename  ->  output extension (prefer 2-char codes)
declare -A MAP=(
  [data_bg.txt]=bg
  [data_en.txt]=en
  [data_fil.txt]=tl      # Filipino; "tl" is the ISO 639-1 code
  [data_hi.txt]=hi
  [data_id.txt]=id
  [data_ja.txt]=ja
  [data_khm.txt]=kh
  [data_lo.txt]=lo
  [data_ms.txt]=ms
  [data_my.txt]=my
  [data_th.txt]=th
  [data_vi.txt]=vi
  [data_zh.txt]=zh
)

for f in "${!MAP[@]}"; do
  lang="${MAP[$f]}"
  python3 split_corpus.py "$INDIR/$f" "$OUTDIR" "$lang" --seed "$SEED"
done

echo
echo "== Sanity check: line counts =="
wc -l "$OUTDIR"/train.* "$OUTDIR"/valid.* "$OUTDIR"/test.* | sort -k2
```

In [21]:
!ls *.sh

split_all.sh  strip_first_column.sh


```bash
#!/bin/bash
# strip_first_column.sh
#
# Remove the first column (sentence ID like SNT.80188.1) from every file
# in an input directory and write the result into an output directory.
#
# Usage:
#   ./strip_first_column.sh <input_dir> <output_dir>
#
# Example:
#   ./strip_first_column.sh ./split ./split.sent
#
set -euo pipefail

INDIR="${1:-./split}"
OUTDIR="${2:-./split.sent}"

if [ ! -d "$INDIR" ]; then
    echo "Error: input directory '$INDIR' not found." >&2
    exit 1
fi

mkdir -p "$OUTDIR"

# awk: split on TAB if present; otherwise fall back to one-or-more spaces.
# We keep the rest of the line untouched (no re-joining => no accidental
# change of internal spacing).
for f in "$INDIR"/*; do
    [ -f "$f" ] || continue
    name="$(basename "$f")"
    out="$OUTDIR/$name"

    awk '
        {
            # If there is at least one TAB, drop up to the first TAB.
            if (index($0, "\t") > 0) {
                sub(/^[^\t]*\t/, "", $0)
            } else {
                # Otherwise drop up to the first run of spaces.
                sub(/^[^ ]+[ ]+/, "", $0)
            }
            print
        }
    ' "$f" > "$out"

    printf '  %-12s -> %-12s  %s\n' "$name" "$name" "$(wc -l < "$out") lines"
done

echo
echo "Done. Output written to: $OUTDIR"
```

In [35]:
%cd /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206

/mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206


In [26]:
!mv data_en.txt data_en.raw.txt

In [27]:
!mv data_en_tok.txt data_en.txt

## Splitting Train/Valid/Test

In [36]:
!./split_all.sh

[th] total= 20106  train= 14074  valid=  4021  test=  2011  -> ./split/train.th, ./split/valid.th, ./split/test.th
[id] total= 20106  train= 14074  valid=  4021  test=  2011  -> ./split/train.id, ./split/valid.id, ./split/test.id
[ms] total= 20106  train= 14074  valid=  4021  test=  2011  -> ./split/train.ms, ./split/valid.ms, ./split/test.ms
[en] total= 20106  train= 14074  valid=  4021  test=  2011  -> ./split/train.en, ./split/valid.en, ./split/test.en
[zh] total= 20106  train= 14074  valid=  4021  test=  2011  -> ./split/train.zh, ./split/valid.zh, ./split/test.zh
[tl] total= 20106  train= 14074  valid=  4021  test=  2011  -> ./split/train.tl, ./split/valid.tl, ./split/test.tl
[lo] total= 20106  train= 14074  valid=  4021  test=  2011  -> ./split/train.lo, ./split/valid.lo, ./split/test.lo
[vi] total= 20106  train= 14074  valid=  4021  test=  2011  -> ./split/train.vi, ./split/valid.vi, ./split/test.vi
[kh] total= 20106  train= 14074  valid=  4021  test=  2011  -> ./split/train.kh,

In [37]:
%cd split

/mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split


In [38]:
!head -n 3 train*

==> train.bg <==
SNT.80188.1	ফ্রান্সের প্যারিসের পার্ক দি প্রিন্সেস-এ হওয়া ২০০৭-এর রাগবি বিশ্বকাপের পুল সি-তে ইটালি পর্তুগালকে ৩১-৫ গোলে হারিয়েছে।
SNT.80188.6	ততসত্ত্বেও বোর্তোলুচ্চি তার ম্যাচের চতুর্থ পেনাল্টিটি মারেন, এবং তারপরে মাউরো বের্গামাস্কোর ও অ্যান্ড্রিয়া মেসির দ্বিতীয় ট্রাই ইটালির জিত নিশ্চিত করে।
SNT.80188.7	বর্তমানে ৮ পয়েন্ট পেয়ে পুল সি-তে তৃতীয় স্থানে থাকা ইটালিকে ২৯ সেপ্টেম্বর দ্বিতীয় স্থানে থাকা স্কটল্যান্ডের বিরুদ্ধে একটি কঠিন ম্যাচ খেলতে হবে।

==> train.en <==
SNT.80188.1	Italy have defeated Portugal 31-5 in Pool C of the 2007 Rugby World Cup at Parc des Princes, Paris, France.
SNT.80188.6	However Bortolussi scored his fourth penalty of the match, followed by tries from Mauro Bergamasco and a second from Andrea Masi to wrap up the win for Italy.
SNT.80188.7	Currently third in Pool C with eight points, Italy face a tough match against second placed Scotland on 29 September.

==> train.hi <==
SNT.80188.1	2007 में फ़्रांस, पेरिस के पार्क डेस प्रिंसेस में हुए रग्बी विश्व 

In [45]:
%cd /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split

/mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split


In [46]:
!head -n 3 valid*

==> valid.bg <==
SNT.80188.2	অ্যান্ড্রিয়া মেসি চার মিনিটের মাথায় ট্রাইব্রেকারের মাধ্যমে ইটালির পক্ষে একটি গোল স্কোর করেন।
SNT.80188.3	প্রথমার্ধের বেশির ভাগ সময় ধরেই খেলা নিজেদের নিয়ন্ত্রণে রাখা সত্ত্বেও ইটালি বিরতির আগে আর কোনো ট্রাই স্কোর করতে পারেনি, কিন্তু ডেভিড বোর্তোলুচ্চি তিনটি পেনাল্টি মেরে তাদের এগিয়ে থাকাকে ধরে রেখেছিলেন।
SNT.80188.4	পর্তুগাল কখনোই হাল ছেড়ে দেয়নি ও ৩৩ মিনিটের মাথায় ডেভিড পেনালভা একটি ট্রাই স্কোর করেছেন, যা কিনা খেলায় তাদের একমাত্র পয়েন্ট এনে দিয়েছে।

==> valid.en <==
SNT.80188.2	Andrea Masi opened the scoring in the fourth minute with a try for Italy.
SNT.80188.3	Despite controlling the game for much of the first half, Italy could not score any other tries before the interval but David Bortolussi kicked three penalties to extend their lead.
SNT.80188.4	Portugal never gave up and David Penalva scored a try in the 33rd minute, providing their only points of the match.

==> valid.hi <==
SNT.80188.2	एंड्रिया मैसी ने चौथे मिनट में इटली के लिए ट्राई के द्वारा स्कोरि

In [47]:
!head -n 3 test*

==> test.bg <==
SNT.87564.14	এমপি-দের কাছে দেওয়া বিবৃতিতে রুথ কেলি নিশ্চিত করেছেন যে হারিয়ে যাওয়া তথ্যের মধ্যে কোনো ব্যাঙ্কিং-এর তথ্য ছিল না, এমনকি কোনো ব্যক্তির জন্মতারিখও ছিল না।
SNT.37376.19	সেনেটর জো লুডউইগ (লেবার, ক্যুইন্সল্যান্ড) প্রশ্নোত্তর সময়ে ন্যায় ও শুল্ক মন্ত্রী সেনেটর ক্রিস এলিসনকে জিজ্ঞাসা করেছেন যে তিনি গতকাল যে দাবি করেছেন যে জিয়ান সেং-এর আবিষ্কার প্রমাণ করে যে অস্ট্রেলীয় সরকারের কাছে "এই পরিস্থিতিতেও একটি জাহাজ আটক করার মতো আন্তরীক্ষ ও সামুদ্রিক নজরদারি আছে, এবং তা করা হয়েছে", সেই দাবি এখনো সমর্থন করেন কিনা এবং জানতে চেয়েছেন যে অস্ট্রেলিয়ার জলের সীমানায় জাহাজটি ঢুকে পড়ার পরে সেটিকে আটক করতে সরকারের ২ সপ্তাহ লাগল কেন।
SNT.78103.38	রেসের ঘোড়ার সংক্রমণের এটিই প্রথম ঘটনা, যদিও গোটা NSW ও ক্যুইন্সল্যান্ড জুড়ে ডজনের বেশি বিনোদনের ঘোড়া আগেই সংক্রামিত হয়েছে।

==> test.en <==
SNT.87564.14	In her statement to MPs, Ruth Kelly confirmed that no banking information was contained within the lost data, nor were individuals' dates of birth.
SNT.37376.19	Senator Joe Ludwig (Labor, Quee

## Remove Sentence ID

In [48]:
%cd /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206

/mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206


In [49]:
!./strip_first_column.sh

  test.bg      -> test.bg       2011 lines
  test.en      -> test.en       2011 lines
  test.hi      -> test.hi       2011 lines
  test.id      -> test.id       2011 lines
  test.ja      -> test.ja       2011 lines
  test.kh      -> test.kh       2011 lines
  test.lo      -> test.lo       2011 lines
  test.ms      -> test.ms       2011 lines
  test.my      -> test.my       2011 lines
  test.th      -> test.th       2011 lines
  test.tl      -> test.tl       2011 lines
  test.vi      -> test.vi       2011 lines
  test.zh      -> test.zh       2011 lines
  train.bg     -> train.bg      14074 lines
  train.en     -> train.en      14074 lines
  train.hi     -> train.hi      14074 lines
  train.id     -> train.id      14074 lines
  train.ja     -> train.ja      14074 lines
  train.kh     -> train.kh      14074 lines
  train.lo     -> train.lo      14074 lines
  train.ms     -> train.ms      14074 lines
  train.my     -> train.my      14074 lines
  train.th     -> train.th      14074 lines
 

In [50]:
%cd split.sent

/mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent


In [51]:
!ls *

test.bg  test.kh  test.tl   train.hi  train.ms	train.zh  valid.ja  valid.th
test.en  test.lo  test.vi   train.id  train.my	valid.bg  valid.kh  valid.tl
test.hi  test.ms  test.zh   train.ja  train.th	valid.en  valid.lo  valid.vi
test.id  test.my  train.bg  train.kh  train.tl	valid.hi  valid.ms  valid.zh
test.ja  test.th  train.en  train.lo  train.vi	valid.id  valid.my


In [52]:
!head -n 3 *

==> test.bg <==
এমপি-দের কাছে দেওয়া বিবৃতিতে রুথ কেলি নিশ্চিত করেছেন যে হারিয়ে যাওয়া তথ্যের মধ্যে কোনো ব্যাঙ্কিং-এর তথ্য ছিল না, এমনকি কোনো ব্যক্তির জন্মতারিখও ছিল না।
সেনেটর জো লুডউইগ (লেবার, ক্যুইন্সল্যান্ড) প্রশ্নোত্তর সময়ে ন্যায় ও শুল্ক মন্ত্রী সেনেটর ক্রিস এলিসনকে জিজ্ঞাসা করেছেন যে তিনি গতকাল যে দাবি করেছেন যে জিয়ান সেং-এর আবিষ্কার প্রমাণ করে যে অস্ট্রেলীয় সরকারের কাছে "এই পরিস্থিতিতেও একটি জাহাজ আটক করার মতো আন্তরীক্ষ ও সামুদ্রিক নজরদারি আছে, এবং তা করা হয়েছে", সেই দাবি এখনো সমর্থন করেন কিনা এবং জানতে চেয়েছেন যে অস্ট্রেলিয়ার জলের সীমানায় জাহাজটি ঢুকে পড়ার পরে সেটিকে আটক করতে সরকারের ২ সপ্তাহ লাগল কেন।
রেসের ঘোড়ার সংক্রমণের এটিই প্রথম ঘটনা, যদিও গোটা NSW ও ক্যুইন্সল্যান্ড জুড়ে ডজনের বেশি বিনোদনের ঘোড়া আগেই সংক্রামিত হয়েছে।

==> test.en <==
In her statement to MPs, Ruth Kelly confirmed that no banking information was contained within the lost data, nor were individuals' dates of birth.
Senator Joe Ludwig (Labor, Queensland) asked of the Minister for Justice and Customs Senator Chr

## Tokenization for English



In [53]:
!ls /home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/*

/home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/basic-protected-patterns
/home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/deescape-special-chars.perl
/home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/deescape-special-chars-PTB.perl
/home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/delete-long-words.perl
/home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/detokenizer.perl
/home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/escape-special-chars.perl
/home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/lowercase.perl
/home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/normalize-punctuation.perl
/home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/pre-tok-clean.perl
/home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/pre_tokenize_cleaning.py
/home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/pre-tokenizer.perl
/home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/remove-non-printing-char

```perl
#!/usr/bin/env perl
#
# This file is part of moses.  Its use is licensed under the GNU Lesser General
# Public License version 2.1 or, at your option, any later version.

use warnings;

# Sample Tokenizer
### Version 1.1
# written by Pidong Wang, based on the code written by Josh Schroeder and Philipp Koehn
# Version 1.1 updates:
#       (1) add multithreading option "-threads NUM_THREADS" (default is 1);
#       (2) add a timing option "-time" to calculate the average speed of this tokenizer;
#       (3) add an option "-lines NUM_SENTENCES_PER_THREAD" to set the number of lines for each thread (default is 2000), and this option controls the memory amount needed: the larger this number is, the larger memory is required (the higher tokenization speed);
### Version 1.0
# $Id: tokenizer.perl 915 2009-08-10 08:15:49Z philipp $
# written by Josh Schroeder, based on code by Philipp Koehn

binmode(STDIN, ":utf8");
binmode(STDOUT, ":utf8");

use warnings;
use FindBin qw($RealBin);
use strict;
use Time::HiRes;

if  (eval {require Thread;1;}) {
  #module loaded
  Thread->import();
}

my $mydir = "$RealBin/../share/nonbreaking_prefixes";

my %NONBREAKING_PREFIX = ();
my @protected_patterns = ();
my $protected_patterns_file = "";
my $language = "en";
my $QUIET = 0;
my $HELP = 0;
my $AGGRESSIVE = 0;
my $SKIP_XML = 0;
my $TIMING = 0;
my $NUM_THREADS = 1;
my $NUM_SENTENCES_PER_THREAD = 2000;
my $PENN = 0;
my $NO_ESCAPING = 0;
while (@ARGV)
{
        $_ = shift;
        /^-b$/ && ($| = 1, next);
        /^-l$/ && ($language = shift, next);
        /^-q$/ && ($QUIET = 1, next);
        /^-h$/ && ($HELP = 1, next);
        /^-x$/ && ($SKIP_XML = 1, next);
        /^-a$/ && ($AGGRESSIVE = 1, next);
        /^-time$/ && ($TIMING = 1, next);
  # Option to add list of regexps to be protected
  /^-protected/ && ($protected_patterns_file = shift, next);
        /^-threads$/ && ($NUM_THREADS = int(shift), next);
        /^-lines$/ && ($NUM_SENTENCES_PER_THREAD = int(shift), next);
        /^-penn$/ && ($PENN = 1, next);
        /^-no-escape/ && ($NO_ESCAPING = 1, next);
}

# for time calculation
my $start_time;
if ($TIMING)
{
    $start_time = [ Time::HiRes::gettimeofday( ) ];
}

# print help message
if ($HELP)
{
        print "Usage ./tokenizer.perl (-l [en|de|...]) (-threads 4) < textfile > tokenizedfile\n";
        print "Options:\n";
        print "  -q     ... quiet.\n";
        print "  -a     ... aggressive hyphen splitting.\n";
        print "  -b     ... disable Perl buffering.\n";
        print "  -time  ... enable processing time calculation.\n";
        print "  -penn  ... use Penn treebank-like tokenization.\n";
        print "  -protected FILE  ... specify file with patters to be protected in tokenisation.\n";
        print "  -no-escape ... don't perform HTML escaping on apostrophy, quotes, etc.\n";
        exit;
}

if (!$QUIET)
{
        print STDERR "Tokenizer Version 1.1\n";
        print STDERR "Language: $language\n";
        print STDERR "Number of threads: $NUM_THREADS\n";
}

# load the language-specific non-breaking prefix info from files in the directory nonbreaking_prefixes
load_prefixes($language,\%NONBREAKING_PREFIX);

if (scalar(%NONBREAKING_PREFIX) eq 0)
{
        print STDERR "Warning: No known abbreviations for language '$language'\n";
}

# Load protected patterns
if ($protected_patterns_file)
{
  open(PP,$protected_patterns_file) || die "Unable to open $protected_patterns_file";
  while(<PP>) {
    chomp;
    push @protected_patterns, $_;
  }
}

my @batch_sentences = ();
my @thread_list = ();
my $count_sentences = 0;

if ($NUM_THREADS > 1)
{# multi-threading tokenization
    while(<STDIN>)
    {
        $count_sentences = $count_sentences + 1;
        push(@batch_sentences, $_);
        if (scalar(@batch_sentences)>=($NUM_SENTENCES_PER_THREAD*$NUM_THREADS))
        {
            # assign each thread work
            for (my $i=0; $i<$NUM_THREADS; $i++)
            {
                my $start_index = $i*$NUM_SENTENCES_PER_THREAD;
                my $end_index = $start_index+$NUM_SENTENCES_PER_THREAD-1;
                my @subbatch_sentences = @batch_sentences[$start_index..$end_index];
                my $new_thread = new Thread \&tokenize_batch, @subbatch_sentences;
                push(@thread_list, $new_thread);
            }
            foreach (@thread_list)
            {
                my $tokenized_list = $_->join;
                foreach (@$tokenized_list)
                {
                    print $_;
                }
            }
            # reset for the new run
            @thread_list = ();
            @batch_sentences = ();
        }
    }
    # the last batch
    if (scalar(@batch_sentences)>0)
    {
        # assign each thread work
        for (my $i=0; $i<$NUM_THREADS; $i++)
        {
            my $start_index = $i*$NUM_SENTENCES_PER_THREAD;
            if ($start_index >= scalar(@batch_sentences))
            {
                last;
            }
            my $end_index = $start_index+$NUM_SENTENCES_PER_THREAD-1;
            if ($end_index >= scalar(@batch_sentences))
            {
                $end_index = scalar(@batch_sentences)-1;
            }
            my @subbatch_sentences = @batch_sentences[$start_index..$end_index];
            my $new_thread = new Thread \&tokenize_batch, @subbatch_sentences;
            push(@thread_list, $new_thread);
        }
        foreach (@thread_list)
        {
            my $tokenized_list = $_->join;
            foreach (@$tokenized_list)
            {
                print $_;
            }
        }
    }
}
else
{# single thread only
    while(<STDIN>)
    {
        if (($SKIP_XML && /^<.+>$/) || /^\s*$/)
        {
            #don't try to tokenize XML/HTML tag lines
            print $_;
        }
        else
        {
            print &tokenize($_);
        }
    }
}

if ($TIMING)
{
    my $duration = Time::HiRes::tv_interval( $start_time );
    print STDERR ("TOTAL EXECUTION TIME: ".$duration."\n");
    print STDERR ("TOKENIZATION SPEED: ".($duration/$count_sentences*1000)." milliseconds/line\n");
}

#####################################################################################
# subroutines afterward

# tokenize a batch of texts saved in an array
# input: an array containing a batch of texts
# return: another array containing a batch of tokenized texts for the input array
sub tokenize_batch
{
    my(@text_list) = @_;
    my(@tokenized_list) = ();
    foreach (@text_list)
    {
        if (($SKIP_XML && /^<.+>$/) || /^\s*$/)
        {
            #don't try to tokenize XML/HTML tag lines
            push(@tokenized_list, $_);
        }
        else
        {
            push(@tokenized_list, &tokenize($_));
        }
    }
    return \@tokenized_list;
}

# the actual tokenize function which tokenizes one input string
# input: one string
# return: the tokenized string for the input string
sub tokenize
{
    my($text) = @_;

    if ($PENN) {
      return tokenize_penn($text);
    }

    chomp($text);
    $text = " $text ";

    # remove ASCII junk
    $text =~ s/\s+/ /g;
    $text =~ s/[\000-\037]//g;

    # Find protected patterns
    my @protected = ();
    foreach my $protected_pattern (@protected_patterns) {
      my $t = $text;
      while ($t =~ /(?<PATTERN>$protected_pattern)(?<TAIL>.*)$/) {
        push @protected, $+{PATTERN};
        $t = $+{TAIL};
      }
    }

    for (my $i = 0; $i < scalar(@protected); ++$i) {
      my $subst = sprintf("THISISPROTECTED%.3d", $i);
      $text =~ s,\Q$protected[$i], $subst ,g;
    }
    $text =~ s/ +/ /g;
    $text =~ s/^ //g;
    $text =~ s/ $//g;

    # seperate out all "other" special characters
    $text =~ s/([^\p{IsAlnum}\s\.\'\`\,\-])/ $1 /g;

    # aggressive hyphen splitting
    if ($AGGRESSIVE)
    {
        $text =~ s/([\p{IsAlnum}])\-(?=[\p{IsAlnum}])/$1 \@-\@ /g;
    }

    #multi-dots stay together
    $text =~ s/\.([\.]+)/ DOTMULTI$1/g;
    while($text =~ /DOTMULTI\./)
    {
        $text =~ s/DOTMULTI\.([^\.])/DOTDOTMULTI $1/g;
        $text =~ s/DOTMULTI\./DOTDOTMULTI/g;
    }

    # seperate out "," except if within numbers (5,300)
    #$text =~ s/([^\p{IsN}])[,]([^\p{IsN}])/$1 , $2/g;

    # separate out "," except if within numbers (5,300)
    # previous "global" application skips some:  A,B,C,D,E > A , B,C , D,E
    # first application uses up B so rule can't see B,C
    # two-step version here may create extra spaces but these are removed later
    # will also space digit,letter or letter,digit forms (redundant with next section)
    $text =~ s/([^\p{IsN}])[,]/$1 , /g;
    $text =~ s/[,]([^\p{IsN}])/ , $1/g;

    # separate "," after a number if it's the end of a sentence
    $text =~ s/([\p{IsN}])[,]$/$1 ,/g;

    # separate , pre and post number
    #$text =~ s/([\p{IsN}])[,]([^\p{IsN}])/$1 , $2/g;
    #$text =~ s/([^\p{IsN}])[,]([\p{IsN}])/$1 , $2/g;

    # turn `into '
    #$text =~ s/\`/\'/g;

    #turn '' into "
    #$text =~ s/\'\'/ \" /g;

    if ($language eq "en")
    {
        #split contractions right
        $text =~ s/([^\p{IsAlpha}])[']([^\p{IsAlpha}])/$1 ' $2/g;
        $text =~ s/([^\p{IsAlpha}\p{IsN}])[']([\p{IsAlpha}])/$1 ' $2/g;
        $text =~ s/([\p{IsAlpha}])[']([^\p{IsAlpha}])/$1 ' $2/g;
        $text =~ s/([\p{IsAlpha}])[']([\p{IsAlpha}])/$1 '$2/g;
        #special case for "1990's"
        $text =~ s/([\p{IsN}])[']([s])/$1 '$2/g;
    }
    elsif (($language eq "fr") or ($language eq "it") or ($language eq "ga"))
    {
        #split contractions left
        $text =~ s/([^\p{IsAlpha}])[']([^\p{IsAlpha}])/$1 ' $2/g;
        $text =~ s/([^\p{IsAlpha}])[']([\p{IsAlpha}])/$1 ' $2/g;
        $text =~ s/([\p{IsAlpha}])[']([^\p{IsAlpha}])/$1 ' $2/g;
        $text =~ s/([\p{IsAlpha}])[']([\p{IsAlpha}])/$1' $2/g;
    }
    else
    {
        $text =~ s/\'/ \' /g;
    }

    #word token method
    my @words = split(/\s/,$text);
    $text = "";
    for (my $i=0;$i<(scalar(@words));$i++)
    {
        my $word = $words[$i];
        if ( $word =~ /^(\S+)\.$/)
        {
            my $pre = $1;
            if (($pre =~ /\./ && $pre =~ /\p{IsAlpha}/) || ($NONBREAKING_PREFIX{$pre} && $NONBREAKING_PREFIX{$pre}==1) || ($i<scalar(@words)-1 && ($words[$i+1] =~ /^[\p{IsLower}]/)))
            {
                #no change
                        }
            elsif (($NONBREAKING_PREFIX{$pre} && $NONBREAKING_PREFIX{$pre}==2) && ($i<scalar(@words)-1 && ($words[$i+1] =~ /^[0-9]+/)))
            {
                #no change
            }
            else
            {
                $word = $pre." .";
            }
        }
        $text .= $word." ";
    }

    # clean up extraneous spaces
    $text =~ s/ +/ /g;
    $text =~ s/^ //g;
    $text =~ s/ $//g;

    # .' at end of sentence is missed
    $text =~ s/\.\' ?$/ . ' /;

    # restore protected
    for (my $i = 0; $i < scalar(@protected); ++$i) {
      my $subst = sprintf("THISISPROTECTED%.3d", $i);
      $text =~ s/$subst/$protected[$i]/g;
    }

    #restore multi-dots
    while($text =~ /DOTDOTMULTI/)
    {
        $text =~ s/DOTDOTMULTI/DOTMULTI./g;
    }
    $text =~ s/DOTMULTI/./g;

    #escape special chars
    if (!$NO_ESCAPING)
      {
        $text =~ s/\&/\&amp;/g;   # escape escape
        $text =~ s/\|/\&#124;/g;  # factor separator
        $text =~ s/\</\&lt;/g;    # xml
        $text =~ s/\>/\&gt;/g;    # xml
        $text =~ s/\'/\&apos;/g;  # xml
        $text =~ s/\"/\&quot;/g;  # xml
        $text =~ s/\[/\&#91;/g;   # syntax non-terminal
        $text =~ s/\]/\&#93;/g;   # syntax non-terminal
      }

    #ensure final line break
    $text .= "\n" unless $text =~ /\n$/;

    return $text;
}

sub tokenize_penn
{
    # Improved compatibility with Penn Treebank tokenization.  Useful if
    # the text is to later be parsed with a PTB-trained parser.
    #
    # Adapted from Robert MacIntyre's sed script:
    #   http://www.cis.upenn.edu/~treebank/tokenizer.sed

    my($text) = @_;
    chomp($text);

    # remove ASCII junk
    $text =~ s/\s+/ /g;
    $text =~ s/[\000-\037]//g;

    # attempt to get correct directional quotes
    $text =~ s/^``/`` /g;
    $text =~ s/^"/`` /g;
    $text =~ s/^`([^`])/` $1/g;
    $text =~ s/^'/`  /g;
    $text =~ s/([ ([{<])"/$1 `` /g;
    $text =~ s/([ ([{<])``/$1 `` /g;
    $text =~ s/([ ([{<])`([^`])/$1 ` $2/g;
    $text =~ s/([ ([{<])'/$1 ` /g;
    # close quotes handled at end

    $text =~ s=\.\.\.= _ELLIPSIS_ =g;

    # separate out "," except if within numbers (5,300)
    $text =~ s/([^\p{IsN}])[,]([^\p{IsN}])/$1 , $2/g;
    # separate , pre and post number
    $text =~ s/([\p{IsN}])[,]([^\p{IsN}])/$1 , $2/g;
    $text =~ s/([^\p{IsN}])[,]([\p{IsN}])/$1 , $2/g;

    #$text =~ s=([;:@#\$%&\p{IsSc}])= $1 =g;
$text =~ s=([;:@#\$%&\p{IsSc}\p{IsSo}])= $1 =g;

    # Separate out intra-token slashes.  PTB tokenization doesn't do this, so
    # the tokens should be merged prior to parsing with a PTB-trained parser
    # (see syntax-hyphen-splitting.perl).
    $text =~ s/([\p{IsAlnum}])\/([\p{IsAlnum}])/$1 \@\/\@ $2/g;

    # Assume sentence tokenization has been done first, so split FINAL periods
    # only.
    $text =~ s=([^.])([.])([\]\)}>"']*) ?$=$1 $2$3 =g;
    # however, we may as well split ALL question marks and exclamation points,
    # since they shouldn't have the abbrev.-marker ambiguity problem
    $text =~ s=([?!])= $1 =g;

    # parentheses, brackets, etc.
    $text =~ s=([\]\[\(\){}<>])= $1 =g;
    $text =~ s/\(/-LRB-/g;
    $text =~ s/\)/-RRB-/g;
    $text =~ s/\[/-LSB-/g;
    $text =~ s/\]/-RSB-/g;
    $text =~ s/{/-LCB-/g;
    $text =~ s/}/-RCB-/g;

    $text =~ s=--= -- =g;

    # First off, add a space to the beginning and end of each line, to reduce
    # necessary number of regexps.
    $text =~ s=$= =;
    $text =~ s=^= =;

    $text =~ s="= '' =g;
    # possessive or close-single-quote
    $text =~ s=([^'])' =$1 ' =g;
    # as in it's, I'm, we'd
    $text =~ s='([sSmMdD]) = '$1 =g;
    $text =~ s='ll = 'll =g;
    $text =~ s='re = 're =g;
    $text =~ s='ve = 've =g;
    $text =~ s=n't = n't =g;
    $text =~ s='LL = 'LL =g;
    $text =~ s='RE = 'RE =g;
    $text =~ s='VE = 'VE =g;
    $text =~ s=N'T = N'T =g;

    $text =~ s= ([Cc])annot = $1an not =g;
    $text =~ s= ([Dd])'ye = $1' ye =g;
    $text =~ s= ([Gg])imme = $1im me =g;
    $text =~ s= ([Gg])onna = $1on na =g;
    $text =~ s= ([Gg])otta = $1ot ta =g;
    $text =~ s= ([Ll])emme = $1em me =g;
    $text =~ s= ([Mm])ore'n = $1ore 'n =g;
    $text =~ s= '([Tt])is = '$1 is =g;
    $text =~ s= '([Tt])was = '$1 was =g;
    $text =~ s= ([Ww])anna = $1an na =g;

    #word token method
    my @words = split(/\s/,$text);
    $text = "";
    for (my $i=0;$i<(scalar(@words));$i++)
    {
        my $word = $words[$i];
        if ( $word =~ /^(\S+)\.$/)
        {
            my $pre = $1;
            if (($pre =~ /\./ && $pre =~ /\p{IsAlpha}/) || ($NONBREAKING_PREFIX{$pre} && $NONBREAKING_PREFIX{$pre}==1) || ($i<scalar(@words)-1 && ($words[$i+1] =~ /^[\p{IsLower}]/)))
            {
                #no change
            }
            elsif (($NONBREAKING_PREFIX{$pre} && $NONBREAKING_PREFIX{$pre}==2) && ($i<scalar(@words)-1 && ($words[$i+1] =~ /^[0-9]+/)))
            {
                #no change
            }
            else
            {
                $word = $pre." .";
            }
        }
        $text .= $word." ";
    }

    # restore ellipses
    $text =~ s=_ELLIPSIS_=\.\.\.=g;

    # clean out extra spaces
    $text =~ s=  *= =g;
    $text =~ s=^ *==g;
    $text =~ s= *$==g;

    #escape special chars
    $text =~ s/\&/\&amp;/g;   # escape escape
    $text =~ s/\|/\&#124;/g;  # factor separator
    $text =~ s/\</\&lt;/g;    # xml
    $text =~ s/\>/\&gt;/g;    # xml
    $text =~ s/\'/\&apos;/g;  # xml
    $text =~ s/\"/\&quot;/g;  # xml
    $text =~ s/\[/\&#91;/g;   # syntax non-terminal
    $text =~ s/\]/\&#93;/g;   # syntax non-terminal

    #ensure final line break
    $text .= "\n" unless $text =~ /\n$/;

    return $text;
}

sub load_prefixes
{
    my ($language, $PREFIX_REF) = @_;

    my $prefixfile = "$mydir/nonbreaking_prefix.$language";

    #default back to English if we don't have a language-specific prefix file
    if (!(-e $prefixfile))
    {
        $prefixfile = "$mydir/nonbreaking_prefix.en";
        print STDERR "WARNING: No known abbreviations for language '$language', attempting fall-back to English version...\n";
        die ("ERROR: No abbreviations files found in $mydir\n") unless (-e $prefixfile);
    }

    if (-e "$prefixfile")
    {
        open(PREFIX, "<:utf8", "$prefixfile");
        while (<PREFIX>)
        {
            my $item = $_;
            chomp($item);
            if (($item) && (substr($item,0,1) ne "#"))
            {
                if ($item =~ /(.*)[\s]+(\#NUMERIC_ONLY\#)/)
                {
                    $PREFIX_REF->{$1} = 2;
                }
                else
                {
                    $PREFIX_REF->{$item} = 1;
                }
            }
        }
        close(PREFIX);
    }
}
```

In [58]:
!time perl /home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/tokenizer.perl -l en -threads 10 < ./train.en > train.en.tok

Tokenizer Version 1.1
Language: en
Number of threads: 10

real	0m0.224s
user	0m0.689s
sys	0m0.050s


In [59]:
!head ./train.en.tok

Italy have defeated Portugal 31-5 in Pool C of the 2007 Rugby World Cup at Parc des Princes , Paris , France .
However Bortolussi scored his fourth penalty of the match , followed by tries from Mauro Bergamasco and a second from Andrea Masi to wrap up the win for Italy .
Currently third in Pool C with eight points , Italy face a tough match against second placed Scotland on 29 September .
New Zealand lead the group with ten points , ahead of Scotland on points difference .
Portugal are bottom of the group with no points , behind Romania with one .
Some personal details of 3 million British learner drivers who had applied for the &apos; theory test &apos; component of their Driving licence have been lost in Iowa , in the USA .
The data was lost on a hard drive that was owned by Pearson Driving Assessments Ltd , a private contractor to the UK driving standards agency .
It is the first major loss of data in the UK since information on 25 million people was lost by HM Revenue in October .


In [60]:
!time perl /home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/tokenizer.perl -l en -threads 10 < ./valid.en > valid.en.tok

Tokenizer Version 1.1
Language: en
Number of threads: 10

real	0m0.266s
user	0m0.446s
sys	0m0.018s


In [61]:
!head ./valid.en.tok

Andrea Masi opened the scoring in the fourth minute with a try for Italy .
Despite controlling the game for much of the first half , Italy could not score any other tries before the interval but David Bortolussi kicked three penalties to extend their lead .
Portugal never gave up and David Penalva scored a try in the 33rd minute , providing their only points of the match .
Italy led 16-5 at half time but were matched by Portugal for much of the second half .
Details were given by the UK Transport Secretary , Ruth Kelly , in the House of Commons at 1730 UTC yesterday .
Senator Ellison responded that at the vessel &apos;s first sighting on March 8 &quot; it was not doing anything illegal &quot; , and criticised the Opposition &apos;s claim that the ship was drifting for seventeen days , saying that &quot; it was not stated as having drifted for the whole of the 17 days &quot; .
Senator David Johnston ( Liberal , Western Australia ) went on to debate that there are no shipping lanes in th

In [62]:
!time perl /home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/tokenizer/tokenizer.perl -l en -threads 10 < ./test.en > test.en.tok

Tokenizer Version 1.1
Language: en
Number of threads: 10

real	0m0.293s
user	0m0.282s
sys	0m0.019s


In [63]:
!head test.en.tok

In her statement to MPs , Ruth Kelly confirmed that no banking information was contained within the lost data , nor were individuals &apos; dates of birth .
Senator Joe Ludwig ( Labor , Queensland ) asked of the Minister for Justice and Customs Senator Chris Ellison in Question Time whether he stood by his claim that he made yesterday that the discovery of the Jian Seng demonstrated that the Australian Government had in place &quot; aerial and maritime surveillance to intercept a vessel in these circumstances , and that was done &quot; , and asked why it took the Government two weeks to intercept the vessel after it entered Australian waters .
The cases are the first infections of race horses , despite infecting dozens of recreational horses across NSW and Queensland .
City officials in seven United States cities were warned today about a possible attack on NFL stadiums using &quot; dirty bombs . &quot;
In response to the threat , several teams are scheduled to sweep some stadiums with

In [64]:
!wc *.en

   2011   46475  282326 test.en
  14074  319738 1946810 train.en
   4021   92018  560890 valid.en
  20106  458231 2790026 total


In [65]:
!wc *.en.tok

   2011   53070  296234 test.en.tok
  14074  365216 2044735 train.en.tok
   4021  105007  589014 valid.en.tok
  20106  523293 2929983 total


## Tokenization/Segmentation for Myanmar Language

In [67]:
%cd /mnt/disk1/ye/exp/alt-tran/

/mnt/disk1/ye/exp/alt-tran


In [68]:
!git clone https://github.com/ye-kyaw-thu/oppaWord

Cloning into 'oppaWord'...
remote: Enumerating objects: 449, done.
remote: Counting objects: 100% (75/75), done.
remote: Compressing objects: 100% (75/75), done.
remote: Total 449 (delta 44), reused 0 (delta 0), pack-reused 374 (from 1)
Receiving objects: 100% (449/449), 288.83 MiB | 17.54 MiB/s, done.
Resolving deltas: 100% (242/242), done.


In [70]:
%cd oppaWord

/mnt/disk1/ye/exp/alt-tran/oppaWord


In [71]:
!7z x data/5gramLM.7z.001


7-Zip 23.01 (x64) : Copyright (c) 1999-2023 Igor Pavlov : 2023-06-20
 64-bit locale=en_US.UTF-8 Threads:24 OPEN_MAX:4096

Scanning the drive for archives:
1 file, 104857600 bytes (100 MiB)

Extracting archive: data/5gramLM.7z.001
--
Path = data/5gramLM.7z.001
Type = Split
Physical Size = 104857600
Volumes = 3
Total Physical Size = 296943811
----
Path = 5gramLM.7z
Size = 296943811
--
Path = 5gramLM.7z
Type = 7z
Physical Size = 296943811
Headers Size = 154
Method = LZMA2:24
Solid = -
Blocks = 1

Everything is Ok

Size:       411659037
Compressed: 296943811


In [73]:
!time python oppa_word.py \
  --input /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/train.my \
  --dict data/myg2p_mypos.dict \
  --arpa ./myMono_clean_syl.trie.bin \
  --use-bimm-fallback \
  --bimm-boost 150 \
  --space-remove-mode "my_not_num" > /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/train.my.tok


real	0m4.602s
user	0m4.521s
sys	0m0.071s


In [74]:
!head /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/train.my.tok

ပြင်သစ် နိုင်ငံ ပါရီ မြို့ ပါ့ ဒက်စ် ပ ရင့် စက်၌ ၂၀၀၇ ခုနှစ် ရပ် ဘီ ကမ္ဘာ့ ဖလား တွင် အီတလီ သည် ပေါ်တူဂီ ကို ၃၁-၅ ဂိုး ဖြင့် ရေကူးကန် စီ တွင် ရှုံးနိမ့် သွား ပါ သည် ။
သို့သော်လည်း ဘော် တို လပ် စီ သည် မော် ရို ဘာ ဂမ် မက် စ ကို၏ ကြိုးစားမှု များ နှင့် အန် ဒ ရီ ယာမာ စီ၏ ဒုတိယ ဂိုး နောက်ပိုင်း အီတလီ အတွက် အနိုင်ရ ရှိ စေရန် သူ၏ စတုတ္ထ မြောက် ကစား ပွဲ၏ ပယ် နယ်လ် တီ ကို သွင်းယူ ပေး ခဲ့ သည် ။
လောလောဆယ် ရေကူးကန် စီ၌ ရှစ် မှတ် နှင့် တတိယ ဖြစ် နေ သော အီတလီ သည် စက်တင် ဘာ ၂၉ တွင် ဒုတိယ ရရှိ ထား သော စကော့တလန် နှင့် ခက်ခဲ သော ယှဉ်ပြိုင် မှု တစ် ခု ကိုရင် ဆိုင် နေ ရ သည် ။
နယူးဇီလန် သည် ရမှတ် ကွာခြားချက် အပေါ် စကော့ တ လန်၏ ရှေ့ မှ ရမှတ် ဆယ် မှတ် နှင့်အတူ အုပ်စု ကိုဦး ဆောင် နေ သည် ။
ပေါ်တူဂီ သည် တစ် မှတ် ရ သော ရို မေး နီး ယား နောက် တွင် အမှတ်မရှိ ဘဲ အုပ် စု၏ အောက်ဆုံး တွင် ရှိ သည် ။
သူ တို့ ယာဉ်မောင်း လိုင် စင်၏ သီအိုရီ စာမေးပွဲ အပိုင်း အတွက် လျှောက်ထား ခဲ့ သော ဗြိတိသျှ သင်ယူ သူ ယာဉ် မောင်း ၃ သန်း၏ ကိုယ် ရေး အသေးစိတ် များ အချို့ အမေရိကန် ပြည်ထောင်စု ရှိ အိုင် အို ဝါ တွင် ပျောက်ဆုံး သွား ခဲ့ သည် ။
အဆိုပါ 

In [75]:
!time python oppa_word.py \
  --input /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/valid.my \
  --dict data/myg2p_mypos.dict \
  --arpa ./myMono_clean_syl.trie.bin \
  --use-bimm-fallback \
  --bimm-boost 150 \
  --space-remove-mode "my_not_num" > /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/valid.my.tok


real	0m1.453s
user	0m1.388s
sys	0m0.064s


In [76]:
!head /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/valid.my.tok

အန် ဒ ရီ ယာမာ စီ သည် အီတလီ အတွက် စမ်းသပ် မှု တစ် ခု အဖြစ် စတုတ္ထ မိနစ် တွင် အမှတ်ပေး ခြင်း ကို ဖွင့်လှစ် ပေး ခဲ့ သည် ။
ပထမ တစ် ဝက်၏ တော်တော် များများ အတွက် ကစားပွဲ ကို ထိန်းချုပ် ခြင်း မရှိ ခဲ့ လျှင် အီတလီ သည် ပွဲ နား ချိန် မ တိုင် မှီ အခြား မည် သည့် ကြိုးစားမှု များ ကို မှ အမှတ် မရ နိုင် ပေ မယ့် ဒေးဗစ် ဘော် တို လပ် စီ သည် သူ တို့၏ဦး ဆောင် မှု ကို အဓွန့်ရှည် စေရန် ပယ် နယ်လ် တီ ၃ ဂိုးသွင်း ပေး ခဲ့ သည် ။
ပေါ်တူဂီ သည် ဘယ် သောအခါ မှ စွန့်လွှတ် မှု မရှိ ခဲ့ ပဲ ဒေးဗစ် ပန် နယ် ဗါ သည် ၃၃ မိနစ် တွင် သူ တို့၏ ကစားပွဲ ရဲ့ အမှတ် များ ကို သာ အထောက်အကူ ပြု သည့် ကြိုးစားမှု တစ် ခုမှ အမှတ်ရ ခဲ့ ပါ သည် ။
အီတလီ သည် ပထမ ပိုင်း၌ ၁၆-၅ ဖြင့်ဦး ဆောင် ခဲ့ သော်လည်း ပေါ်တူဂီ မှ ဒုတိယပိုင်း တော်တော် များများ တွင် ဂိုး အရေအတွက် ညီ စေ ခဲ့ သည် ။
အသေးစိတ် များ ကို မနေ့ က ၁၇၃၀ ယူ တီ စီ၌ အောက်လွှတ်တော် ရှိ ဗြိတိန် နိုင်ငံ ပို့ဆောင် ရေး အတွင်းရေးမှူး ရုသ် ကယ် လီ က ပေး ခဲ့ သည် ။
အထက်လွှတ်တော် အမတ် အယ် လစ် ဆန် သည် မတ် လ ၈ ရက်နေ့ တွင်င်္သ ဘော၏ ပထမဆုံး ရှာဖွေ ခြင်း၌ “ တရား မ၀င် သည့် မည် သည့် အရာ မှ မ လုပ် ခဲ့ ပါ ” ဟု တုန့်

In [77]:
!time python oppa_word.py \
  --input /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/test.my \
  --dict data/myg2p_mypos.dict \
  --arpa ./myMono_clean_syl.trie.bin \
  --use-bimm-fallback \
  --bimm-boost 150 \
  --space-remove-mode "my_not_num" > /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/test.my.tok


real	0m0.877s
user	0m0.796s
sys	0m0.081s


In [78]:
!head /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/test.my.tok

လွှတ်တော် အမတ် များ သို့ သူ မ၏ ကြေငြာချက် တွင် ရုသ် ကယ် လီ က ပျောက်ဆုံး ခဲ့ သော အချက်အလက် များ အတွင်း မည် သည့် ဘဏ် လုပ်ငန်း သတင်း အချက်အလက် မှ မ ပါ၀င် သကဲ့သို့ တစ်ဦး ချင်း စီ၏ မွေးသက္ကရာဇ် များ လည်း မ ပါ၀င် ကြောင်း အတည်ပြု သွား ခဲ့ သည် ။
အထက်လွှတ်တော် အမတ် ဂျိုး လူး၀တ် ( အလုပ်သမား ရေးရာ ၊ ကွင်းစ လန် ) က တရား ရေး ဝန်ကြီး နှင့် အကောက်ခွန် အထက်လွှတ်တော် အမတ် ချ ရစ် အယ် လစ် ဆန် အား သြစတြေးလျ အစိုးရ ပိုင်နက် တွင် “ ဤ အခြေအနေ များ တွင် သင်္ဘော တစ်စီး အား ဖမ်းဆီး ရန်ေ၀ ဟင် နှင့် ရေကြောင်း စောင့်ကြည့် မှု နှင့် ထို အလုပ် မှာ ပြီးဆုံး ခဲ့ ကြောင်း ” ကို သရုပ်ပြ ခဲ့ သည့် ဂျီ အန်း ဆန် ရှာဖွေ မှု နှင့် ပတ် သက်၍ မနေ့ က သူ ပြုလုပ် ခဲ့ သည် ဆို သော သူ၏ တောင်းဆိုချက် အတွက် သူ ရပ်တည် ခြင်း ရှိ မရှိ ကို မေးခွန်း မေး ချိန် တွင် မေး ခဲ့ ပြီး အစိုးရ က သြစတြေးလျ ရေကြောင်း သို့ ၀င် ရောက် ခဲ့ ပြီးနောက် သင်္ဘော ကို ဖမ်းဆီး ရန် ကိစ္စ အား မည် သည့် အတွက်ကြောင့် အချိန် နှစ် ပတ် ယူ ခဲ့ သည် ကို မေး ခဲ့ သည် ။
အန် အက်(စ်) ဒ ဗ လျူ နှင့် ကွင်း(စ်) လန်း(ဒ်) တလျှောက် မှ အပန်းဖြေ ရာ သုံး သော မြင်း များ ဒါဇင် များ စွာ ကူးစက် 

## Syllable Segmentation for Myanmar Language

In [80]:
%cd /mnt/disk1/ye/exp/alt-tran

/mnt/disk1/ye/exp/alt-tran


In [81]:
!wget https://raw.githubusercontent.com/ye-kyaw-thu/sylbreak4all/refs/heads/main/python/sylbreak4all.py

--2026-09-26 13:16:50--  https://raw.githubusercontent.com/ye-kyaw-thu/sylbreak4all/refs/heads/main/python/sylbreak4all.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.108.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 4796 (4.7K) [text/plain]
Saving to: ‘sylbreak4all.py’

sylbreak4all.py     100%[===================>]   4.68K  --.-KB/s    in 0s      

2026-09-26 13:16:51 (135 MB/s) - ‘sylbreak4all.py’ saved [4796/4796]



```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""sylbreak4all: regex-based syllable breaking for nine Myanmar languages."""

from __future__ import annotations

import re
import sys
from pathlib import Path
from typing import TextIO

MY_CONSONANT = r"က-အ"
EN_CHAR = r"a-zA-Z0-9"
SS_SYMBOL = "္"
ATHAT = "်"
OTHER_CHAR = r"ဣဤဥဦဧဩဪဿ၌၍၎၏၀-၉႐-႙၊။!-/:-@[-`’“”{-~\s…"

SH_CONSONANT = r"ၵၶငၸသၺတထၼပၽၾမယရလဝႁဢၹၷႀၻၿ"
OTHER_SH = r"႟႐-႙၊။!-/:-@[-`{-~\s"
SK_CONSONANT = r"ကခဂဃငစဆဇညတထဒနပဖဘမယရလဝသဟအဧ"
OTHER_SK = r"ဒမၡဧ၀-၉၊။!-/:-@[( -`{-~\s".replace(" ", "")
OTHER_PK = r"ၥၦၡဧ၀-၉၊။!-/:-@[( -`{-~\s".replace(" ", "")
MO_CONSONANT = r"ကခဂဃၚစဆဇၛဉညဋဌဍဎဏတထဒဓနပဖဗဘမယရလဝသဟဠၜအၝ"


def compile_patterns() -> dict[str, re.Pattern[str]]:
    common = rf"((?<!{SS_SYMBOL})[{MY_CONSONANT}](?![{ATHAT}{SS_SYMBOL}])|[{EN_CHAR}{OTHER_CHAR}])"
    return {
        "bm": re.compile(common),
        "rk": re.compile(common),
        "dw": re.compile(common),
        "bk": re.compile(common),
        "po": re.compile(common),
        "sh": re.compile(rf"([{SH_CONSONANT}](?![{ATHAT}])|[{EN_CHAR}{OTHER_SH}])"),
        "sk": re.compile(rf"([{SK_CONSONANT}]|[{EN_CHAR}{OTHER_SK}])"),
        "pk": re.compile(rf"([{MY_CONSONANT}]|[{EN_CHAR}{OTHER_PK}])"),
        "mo": re.compile(rf"((?<!{SS_SYMBOL})[{MO_CONSONANT}](?![{ATHAT}{SS_SYMBOL}])|[{EN_CHAR}{OTHER_CHAR}])"),
    }


def segment_line(line: str, pattern: re.Pattern[str], separator: str) -> str:
    # Match the historical sylbreak4all reference behavior: phrase whitespace is
    # removed before applying the syllable-breaking RE.
    line = re.sub(r"\s+", "", line)
    output = pattern.sub(lambda m: separator + m.group(1), line)
    return re.sub(r"^\s+|\s+$", "", output)


def open_input(path: str | None) -> TextIO:
    if path is None:
        return sys.stdin
    return Path(path).open("r", encoding="utf-8")


HELP_TEXT = """Syllable Breaker Tool
Version: 0.9.0

Supported Languages:
    bm  - Burmese
    bk  - Beik
    dw  - Dawei
    rk  - Rakhine
    mo  - Mon
    po  - PaO
    sh  - Shan
    sk  - Sgaw Kayin
    pk  - Pwo Kayin

Usage: sylbreak4all [-i filename] [-l language] [-s separator] [-p]
    -i | --input-file   Input file name (default: STDIN)
    -s | --separator    Separator (default: '|')
    -l | --language     Language (default: bm for Burmese)
    -p | --print        Print input alongside the output (default: 0)
    -h | --help         Show this help
"""

def print_help(stream: TextIO = sys.stdout) -> None:
    print(HELP_TEXT, end="", file=stream)

def main() -> int:
    input_file = None
    separator = "|"
    language = "bm"
    print_input = False
    args = sys.argv[1:]
    i = 0
    while i < len(args):
        arg = args[i]
        if arg in ("-h", "--help"):
            print_help()
            return 0
        if arg in ("-i", "--input", "--input-file"):
            i += 1
            if i >= len(args):
                print("Missing value for -i/--input-file", file=sys.stderr)
                print_help(sys.stderr)
                return 1
            input_file = args[i]
        elif arg in ("-s", "--separator"):
            i += 1
            if i >= len(args):
                print("Missing value for -s/--separator", file=sys.stderr)
                print_help(sys.stderr)
                return 1
            separator = args[i]
        elif arg in ("-l", "--language", "--lang"):
            i += 1
            if i >= len(args):
                print("Missing value for -l/--language", file=sys.stderr)
                print_help(sys.stderr)
                return 1
            language = args[i]
        elif arg in ("-p", "--print"):
            print_input = True
        else:
            print(f"Unknown argument: {arg}", file=sys.stderr)
            print_help(sys.stderr)
            return 1
        i += 1

    patterns = compile_patterns()
    if language not in patterns:
        print(f"Unsupported language: {language}", file=sys.stderr)
        print_help(sys.stderr)
        return 1

    fh = open_input(input_file)
    try:
        for raw_line in fh:
            line = raw_line.rstrip("\r\n")
            output = segment_line(line, patterns[language], separator)
            if print_input:
                clean = re.sub(r"\s+", "", line)
                print(f"input: {clean}")
                print(f"output: {output}")
            else:
                print(output)
    finally:
        if fh is not sys.stdin:
            fh.close()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
```

In [82]:
!python ./sylbreak4all.py --help

Syllable Breaker Tool
Version: 0.9.0

Supported Languages:
    bm  - Burmese
    bk  - Beik
    dw  - Dawei
    rk  - Rakhine
    mo  - Mon
    po  - PaO
    sh  - Shan
    sk  - Sgaw Kayin
    pk  - Pwo Kayin

Usage: sylbreak4all [-i filename] [-l language] [-s separator] [-p]
    -i | --input-file   Input file name (default: STDIN)
    -s | --separator    Separator (default: '|')
    -l | --language     Language (default: bm for Burmese)
    -p | --print        Print input alongside the output (default: 0)
    -h | --help         Show this help


In [84]:
!python ./sylbreak4all.py -i /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/train.my -l bm -s " " > /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/train.my.syl

In [85]:
!head /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/train.my.syl

ပြင် သစ် နိုင် ငံ ပါ ရီ မြို့ ပါ့ ဒက်စ် ပ ရင့် စက် ၌ ၂ ၀ ၀ ၇ ခု နှစ် ရပ် ဘီ ကမ္ဘာ့ ဖ လား တွင် အီ တ လီ သည် ပေါ် တူ ဂီ ကို ၃ ၁ - ၅ ဂိုး ဖြင့် ရေ ကူး ကန် စီ တွင် ရှုံး နိမ့် သွား ပါ သည် ။
သို့ သော် လည်း ဘော် တို လပ် စီ သည် မော် ရို ဘာ ဂမ် မက် စ ကို ၏ ကြိုး စား မှု များ နှင့် အန် ဒ ရီ ယာ မာ စီ ၏ ဒု တိ ယ ဂိုး နောက် ပိုင်း အီ တ လီ အ တွက် အ နိုင် ရ ရှိ စေ ရန် သူ ၏ စ တုတ္ထ မြောက် က စား ပွဲ ၏ ပယ် နယ်လ် တီ ကို သွင်း ယူ ပေး ခဲ့ သည် ။
လော လော ဆယ် ရေ ကူး ကန် စီ ၌ ရှစ် မှတ် နှင့် တ တိ ယ ဖြစ် နေ သော အီ တ လီ သည် စက် တင် ဘာ ၂ ၉ တွင် ဒု တိ ယ ရ ရှိ ထား သော စ ကော့ တ လန် နှင့် ခက် ခဲ သော ယှဉ် ပြိုင် မှု တစ် ခု ကို ရင် ဆိုင် နေ ရ သည် ။
န ယူး ဇီ လန် သည် ရ မှတ် ကွာ ခြား ချက် အ ပေါ် စ ကော့ တ လန် ၏ ရှေ့ မှ ရ မှတ် ဆယ် မှတ် နှင့် အ တူ အုပ် စု ကို ဦး ဆောင် နေ သည် ။
ပေါ် တူ ဂီ သည် တစ် မှတ် ရ သော ရို မေး နီး ယား နောက် တွင် အ မှတ် မ ရှိ ဘဲ အုပ် စု ၏ အောက် ဆုံး တွင် ရှိ သည် ။
သူ တို့ ယာဉ် မောင်း လိုင် စင် ၏ သီ အို ရီ စာ မေး ပွဲ အ ပိုင်း အ တွက် လျှောက် ထား ခဲ့ သော ဗြိ တိ သျှ သင် ယူ သူ ယာဉ် မောင်း ၃ သန်း ၏ ကိုယ် ရေး အ သ

In [86]:
!python ./sylbreak4all.py -i /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/valid.my -l bm -s " " > /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/valid.my.syl

In [87]:
!head /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/valid.my.syl

အန် ဒ ရီ ယာ မာ စီ သည် အီ တ လီ အ တွက် စမ်း သပ် မှု တစ် ခု အ ဖြစ် စ တုတ္ထ မိ နစ် တွင် အ မှတ် ပေး ခြင်း ကို ဖွင့် လှစ် ပေး ခဲ့ သည် ။
ပ ထ မ တစ် ဝက် ၏ တော် တော် များ များ အ တွက် က စား ပွဲ ကို ထိန်း ချုပ် ခြင်း မ ရှိ ခဲ့ လျှင် အီ တ လီ သည် ပွဲ နား ချိန် မ တိုင် မှီ အ ခြား မည် သည့် ကြိုး စား မှု များ ကို မှ အ မှတ် မ ရ နိုင် ပေ မယ့် ဒေး ဗစ် ဘော် တို လပ် စီ သည် သူ တို့ ၏ ဦး ဆောင် မှု ကို အ ဓွန့် ရှည် စေ ရန် ပယ် နယ်လ် တီ ၃ ဂိုး သွင်း ပေး ခဲ့ သည် ။
ပေါ် တူ ဂီ သည် ဘယ် သော အ ခါ မှ စွန့် လွှတ် မှု မ ရှိ ခဲ့ ပဲ ဒေး ဗစ် ပန် နယ် ဗါ သည် ၃ ၃ မိ နစ် တွင် သူ တို့ ၏ က စား ပွဲ ရဲ့ အ မှတ် များ ကို သာ အ ထောက် အ ကူ ပြု သည့် ကြိုး စား မှု တစ် ခု မှ အ မှတ် ရ ခဲ့ ပါ သည် ။
အီ တ လီ သည် ပ ထ မ ပိုင်း ၌ ၁ ၆ - ၅ ဖြင့် ဦး ဆောင် ခဲ့ သော် လည်း ပေါ် တူ ဂီ မှ ဒု တိ ယ ပိုင်း တော် တော် များ များ တွင် ဂိုး အ ရေ အ တွက် ညီ စေ ခဲ့ သည် ။
အ သေး စိတ် များ ကို မ နေ့ က ၁ ၇ ၃ ၀ ယူ တီ စီ ၌ အောက် လွှတ် တော် ရှိ ဗြိ တိန် နိုင် ငံ ပို့ ဆောင် ရေး အ တွင်း ရေး မှူး ရုသ် ကယ် လီ က ပေး ခဲ့ သည် ။
အ ထက် လွှတ် တော် အ မတ် အယ် လစ် ဆန် သည် မတ် လ ၈ ရက် န

In [88]:
!python ./sylbreak4all.py -i /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/test.my -l bm -s " " > /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/test.my.syl

In [89]:
!head /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent/test.my.syl

လွှတ် တော် အ မတ် များ သို့ သူ မ ၏ ကြေ ငြာ ချက် တွင် ရုသ် ကယ် လီ က ပျောက် ဆုံး ခဲ့ သော အ ချက် အ လက် များ အ တွင်း မည် သည့် ဘဏ် လုပ် ငန်း သ တင်း အ ချက် အ လက် မှ မ ပါ ၀င် သ ကဲ့ သို့ တစ် ဦး ချင်း စီ ၏ မွေး သက္က ရာဇ် များ လည်း မ ပါ ၀င် ကြောင်း အ တည် ပြု သွား ခဲ့ သည် ။
အ ထက် လွှတ် တော် အ မတ် ဂျိုး လူး ၀တ် ( အ လုပ် သ မား ရေး ရာ ၊ ကွင်း စ လန် ) က တ ရား ရေး ဝန် ကြီး နှင့် အ ကောက် ခွန် အ ထက် လွှတ် တော် အ မတ် ချ ရစ် အယ် လစ် ဆန် အား သြ စ တြေး လျ အ စိုး ရ ပိုင် နက် တွင် “ ဤ အ ခြေ အ နေ များ တွင် သင်္ဘော တစ် စီး အား ဖမ်း ဆီး ရန်ေ ၀ ဟင် နှင့် ရေ ကြောင်း စောင့် ကြည့် မှု နှင့် ထို အ လုပ် မှာ ပြီး ဆုံး ခဲ့ ကြောင်း ” ကို သ ရုပ် ပြ ခဲ့ သည့် ဂျီ အန်း ဆန် ရှာ ဖွေ မှု နှင့် ပတ် သက် ၍ မ နေ့ က သူ ပြု လုပ် ခဲ့ သည် ဆို သော သူ ၏ တောင်း ဆို ချက် အ တွက် သူ ရပ် တည် ခြင်း ရှိ မ ရှိ ကို မေး ခွန်း မေး ချိန် တွင် မေး ခဲ့ ပြီး အ စိုး ရ က သြ စ တြေး လျ ရေ ကြောင်း သို့ ၀င် ရောက် ခဲ့ ပြီး နောက် သင်္ဘော ကို ဖမ်း ဆီး ရန် ကိစ္စ အား မည် သည့် အ တွက် ကြောင့် အ ချိန် နှစ် ပတ် ယူ ခဲ့ သည် ကို မေး ခဲ့ သည် ။
အန် အက် (စ် ) ဒ ဗ လျူ နှင့် 

## Marian NMT Framework

In [90]:
!which marian

/home/ye/tool/marian/build/marian


In [91]:
!grep "marian" /home/ye/.bashrc

export PATH="/usr/bin:/bin:/home/ye/tool/marian/build:$PATH"
#export PATH="/home/ye/tool/marian/build:$PATH"


In [92]:
!marian --help

Marian: Fast Neural Machine Translation in C++
Usage: marian [OPTIONS]

General options:
  -h,--help                             Print this help message and exit
  --version                             Print the version number and exit
  --authors                             Print list of authors and exit
  --cite                                Print citation and exit
  --build-info TEXT                     Print CMake build options and exit. Set to 'all' to print advanced options
  -c,--config VECTOR ...                Configuration file(s). If multiple, later overrides earlier
  -w,--workspace INT=2048               Preallocate arg MB of work space. Negative `--workspace -N` value allocates workspace as total available GPU memory minus N megabytes.
  --log TEXT                            Log training process information to file given by arg
  --log-level TEXT=info                 Set verbosity level of logging: trace, debug, info, warn, err(or), critical, off
  --log-time-zone TEXT  

In [93]:
!marian-vocab --help

Create a vocabulary from text corpora given on STDIN
Usage: marian-vocab [OPTIONS]

Allowed options:
  -h,--help                             Print this help message and exit
  --version                             Print the version number and exit
  -m,--max-size UINT=0                  Generate only UINT most common vocabulary items

Examples:
  ./marian-vocab < text.src > vocab.yml
  cat text.src text.trg | ./marian-vocab > vocab.yml


## Vocab Building 

In [94]:
%pwd

'/mnt/disk1/ye/exp/alt-tran'

In [95]:
%mkdir -p data/vocab

In [96]:
%cd /mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent

/mnt/disk1/ye/exp/alt-tran/ALT-Parallel-Corpus-20191206/split.sent


In [98]:
!cat train.en.tok valid.en.tok test.en.tok | marian-vocab > /mnt/disk1/ye/exp/alt-tran/data/vocab/en_vocab.yml

[2026-09-26 13:44:43] Creating vocabulary...
[2026-09-26 13:44:43] [data] Creating vocabulary stdout from stdin
[2026-09-26 13:44:43] Finished


In [99]:
!head -n 20 /mnt/disk1/ye/exp/alt-tran/data/vocab/en_vocab.yml

</s>: 0
<unk>: 1
the: 2
",": 3
.: 4
of: 5
to: 6
and: 7
in: 8
"&quot;": 9
a: 10
that: 11
The: 12
for: 13
was: 14
is: 15
on: 16
"&apos;s": 17
by: 18
with: 19


In [100]:
!tail -n 20 /mnt/disk1/ye/exp/alt-tran/data/vocab/en_vocab.yml

Ötzi: 34164
Úrsula: 34165
Ürümqi: 34166
à: 34167
émigré: 34168
İstanbul: 34169
İzmir: 34170
Špidla: 34171
μg: 34172
сhampionship: 34173
‎: 34174
‪: 34175
₣: 34176
−: 34177
、: 34178
」: 34179
匯聚智慧: 34180
珠穆朗瑪峰: 34181
集體創新: 34182
ｖ: 34183

In [102]:
!wc /mnt/disk1/ye/exp/alt-tran/data/vocab/en_vocab.yml

 34183  68368 514447 /mnt/disk1/ye/exp/alt-tran/data/vocab/en_vocab.yml


**Building vocab file for Myanmar Word data**

In [103]:
!cat train.my.tok valid.my.tok test.my.tok | marian-vocab > /mnt/disk1/ye/exp/alt-tran/data/vocab/my_vocab.yml

[2026-09-26 13:49:05] Creating vocabulary...
[2026-09-26 13:49:05] [data] Creating vocabulary stdout from stdin
[2026-09-26 13:49:05] Finished


In [104]:
!head -n 20 /mnt/disk1/ye/exp/alt-tran/data/vocab/my_vocab.yml

</s>: 0
<unk>: 1
သည်: 2
ခဲ့: 3
၊: 4
ကို: 5
။: 6
များ: 7
သော: 8
က: 9
တွင်: 10
မှု: 11
တစ်: 12
နှင့်: 13
သူ: 14
ခု: 15
ပြီး: 16
ဖြစ်: 17
ခြင်း: 18
ရှိ: 19


In [105]:
!tail -n 20 /mnt/disk1/ye/exp/alt-tran/data/vocab/my_vocab.yml

၊၁၀: 19632
၊၂၀၀၀: 19633
၊၄၆၄: 19634
၊၌: 19635
၊၌၎င်း: 19636
၊၎င်းဥ: 19637
၊၎င်း၏ဦး: 19638
။.: 19639
။ဤ: 19640
၎င်း၏ဥက္က: 19641
၏ဥ: 19642
“၎င်း: 19643
…..”: 19644
….”: 19645
……..: 19646
€: 19647
、: 19648
・: 19649
匯聚智慧: 19650
集體創新: 19651

In [109]:
!wc /mnt/disk1/ye/exp/alt-tran/data/vocab/my_vocab.yml

 19651  39304 519277 /mnt/disk1/ye/exp/alt-tran/data/vocab/my_vocab.yml


**Building vocab file for Myanmar Syllable data**

In [106]:
!cat train.my.syl valid.my.syl test.my.syl | marian-vocab > /mnt/disk1/ye/exp/alt-tran/data/vocab/my_syl_vocab.yml

[2026-09-26 13:51:53] Creating vocabulary...
[2026-09-26 13:51:53] [data] Creating vocabulary stdout from stdin
[2026-09-26 13:51:53] Finished


In [107]:
!head -n 20 /mnt/disk1/ye/exp/alt-tran/data/vocab/my_syl_vocab.yml

</s>: 0
<unk>: 1
အ: 2
သည်: 3
ခဲ့: 4
ကို: 5
များ: 6
၊: 7
။: 8
က: 9
သော: 10
၏: 11
မှု: 12
သူ: 13
တစ်: 14
တွင်: 15
မ: 16
ပါ: 17
ရှိ: 18
ဖြစ်: 19


In [108]:
!tail -n 20 /mnt/disk1/ye/exp/alt-tran/data/vocab/my_syl_vocab.yml

၌ြ: 5093
၎: 5094
၏င်္သ: 5095
၏ေ: 5096
၏ေြ: 5097
၏ြ: 5098
၏—: 5099
႐စ်: 5100
႐တ်: 5101
႐မ်း: 5102
႐ဴး: 5103
႐ံ: 5104
႐ွန်: 5105
႑: 5106
႒န်း: 5107
႒ပ်: 5108
႒ြ: 5109
႕ေ: 5110
“ေ: 5111
“‘: 5112

In [110]:
!wc /mnt/disk1/ye/exp/alt-tran/data/vocab/my_syl_vocab.yml

  5112  10226 102199 /mnt/disk1/ye/exp/alt-tran/data/vocab/my_syl_vocab.yml


## Move/Prepare Train/Valid/Test Data

In [112]:
!tree /mnt/disk1/ye/exp/alt-tran/data

/mnt/disk1/ye/exp/alt-tran/data
├── syl
│   ├── test.en
│   ├── test.my
│   ├── train.en
│   ├── train.my
│   ├── valid.en
│   └── valid.my
├── vocab
│   ├── en_vocab.yml
│   ├── my_syl_vocab.yml
│   └── my_vocab.yml
└── word
    ├── test.en
    ├── test.my
    ├── train.en
    ├── train.my
    ├── valid.en
    └── valid.my

4 directories, 15 files


In [113]:
!wc /mnt/disk1/ye/exp/alt-tran/data/syl/*.my

    2011   118037  1136378 /mnt/disk1/ye/exp/alt-tran/data/syl/test.my
   14074   815181  7827870 /mnt/disk1/ye/exp/alt-tran/data/syl/train.my
    4021   234217  2251451 /mnt/disk1/ye/exp/alt-tran/data/syl/valid.my
   20106  1167435 11215699 total


In [115]:
!wc /mnt/disk1/ye/exp/alt-tran/data/syl/*.en

   2011   53070  296234 /mnt/disk1/ye/exp/alt-tran/data/syl/test.en
  14074  365216 2044735 /mnt/disk1/ye/exp/alt-tran/data/syl/train.en
   4021  105007  589014 /mnt/disk1/ye/exp/alt-tran/data/syl/valid.en
  20106  523293 2929983 total


In [116]:
!wc /mnt/disk1/ye/exp/alt-tran/data/word/*.my

    2011    78949  1097290 /mnt/disk1/ye/exp/alt-tran/data/word/test.my
   14074   546497  7559186 /mnt/disk1/ye/exp/alt-tran/data/word/train.my
    4021   157101  2174335 /mnt/disk1/ye/exp/alt-tran/data/word/valid.my
   20106   782547 10830811 total


## Bash Shell Script for Transformer NMT Model

```bash
#!/bin/bash

## Written by Ye Kyaw Thu, Language Understanding Lab, AI Mario, Myanmar
## Experiments for AI Engineering Fundamental Class (Batch-2)
## Date: 26 Sept 2026

## Old Notes
#     --mini-batch-fit -w 10000 --maxi-batch 1000 \
#    --mini-batch-fit -w 1000 --maxi-batch 100 \
#     --tied-embeddings-all \
#     --tied-embeddings \
#     --valid-metrics cross-entropy perplexity translation bleu \
#     --transformer-dropout 0.1 --label-smoothing 0.1 \
#     --learn-rate 0.0003 --lr-warmup 16000 --lr-decay-inv-sqrt 16000 --lr-report \
#     --optimizer-params 0.9 0.98 1e-09 --clip-norm 5 \

model_folder="model.tf.syl";
mkdir -p ${model_folder};
data_path="/mnt/disk1/ye/exp/alt-tran/data/syl/";
vocab_path="/mnt/disk1/ye/exp/alt-tran/data/vocab/"
src="en"; tgt="my";

marian \
    --model ${model_folder}/model.npz --type transformer \
    --train-sets ${data_path}/train.${src} ${data_path}/train.${tgt} \
    --max-length 200 \
    --vocabs  ${vocab_path}/en_vocab.yml  ${vocab_path}/my_syl_vocab.yml \
    --mini-batch-fit -w 1000 --maxi-batch 100 \
    --early-stopping 5 \
    --valid-freq 5000 --save-freq 5000 --disp-freq 500 \
    --valid-metrics cross-entropy perplexity bleu \
    --valid-sets ${data_path}/valid.${src} ${data_path}/valid.${tgt} \
    --valid-translation-output ${model_folder}/valid.${src}-${tgt}.output \
    --quiet-translation \
    --valid-mini-batch 64 \
    --beam-size 6 --normalize 0.6 \
    --log ${model_folder}/train.log \
    --valid-log ${model_folder}/valid.log \
    --enc-depth 2 --dec-depth 2 \
    --transformer-heads 8 \
    --transformer-postprocess-emb d \
    --transformer-postprocess dan \
    --transformer-dropout 0.3 --label-smoothing 0.1 \
    --learn-rate 0.0003 --lr-warmup 0 --lr-decay-inv-sqrt 16000 --lr-report \
    --clip-norm 5 \
    --tied-embeddings \
    --devices 0 --sync-sgd --seed 1111 \
    --exponential-smoothing \
    --dump-config > ${model_folder}/${src}-${tgt}.config.yml

time marian -c ${model_folder}/${src}-${tgt}.config.yml 2>&1 | tee ${model_folder}/tf.${src}-${tgt}.log

```

In [119]:
%cd /mnt/disk1/ye/exp/alt-tran

/mnt/disk1/ye/exp/alt-tran


In [120]:
!./tf.syl.sh

[2026-09-26 14:28:29] [marian] Marian v1.12.0 65bf82ff 2023-02-21 09:56:29 -0800
[2026-09-26 14:28:29] [marian] Running on lst-hpc3090 as process 1559687 with command line:
[2026-09-26 14:28:29] [marian] marian -c model.tf.syl/en-my.config.yml
[2026-09-26 14:28:29] [config] after: 0e
[2026-09-26 14:28:29] [config] after-batches: 0
[2026-09-26 14:28:29] [config] after-epochs: 0
[2026-09-26 14:28:29] [config] all-caps-every: 0
[2026-09-26 14:28:29] [config] allow-unk: false
[2026-09-26 14:28:29] [config] authors: false
[2026-09-26 14:28:29] [config] beam-size: 6
[2026-09-26 14:28:29] [config] bert-class-symbol: "[CLS]"
[2026-09-26 14:28:29] [config] bert-mask-symbol: "[MASK]"
[2026-09-26 14:28:29] [config] bert-masking-fraction: 0.15
[2026-09-26 14:28:29] [config] bert-sep-symbol: "[SEP]"
[2026-09-26 14:28:29] [config] bert-train-type-embeddings: true
[2026-09-26 14:28:29] [config] bert-type-vocab-size: 2
[2026-09-26 14:28:29] [config] build-info: ""
[2026-09-26 14:28:29] [config] check-

## Updated Shell Script for Transformer Model (Syllable)

```
model_folder="model.tf.es10.syl";
    --early-stopping 10  
```

In [122]:
! ./tf.syl.es10.sh

[2026-09-26 14:45:10] [marian] Marian v1.12.0 65bf82ff 2023-02-21 09:56:29 -0800
[2026-09-26 14:45:10] [marian] Running on lst-hpc3090 as process 1563310 with command line:
[2026-09-26 14:45:10] [marian] marian -c model.tf.es10.syl/en-my.config.yml
[2026-09-26 14:45:10] [config] after: 0e
[2026-09-26 14:45:10] [config] after-batches: 0
[2026-09-26 14:45:10] [config] after-epochs: 0
[2026-09-26 14:45:10] [config] all-caps-every: 0
[2026-09-26 14:45:10] [config] allow-unk: false
[2026-09-26 14:45:10] [config] authors: false
[2026-09-26 14:45:10] [config] beam-size: 6
[2026-09-26 14:45:10] [config] bert-class-symbol: "[CLS]"
[2026-09-26 14:45:10] [config] bert-mask-symbol: "[MASK]"
[2026-09-26 14:45:10] [config] bert-masking-fraction: 0.15
[2026-09-26 14:45:10] [config] bert-sep-symbol: "[SEP]"
[2026-09-26 14:45:10] [config] bert-train-type-embeddings: true
[2026-09-26 14:45:10] [config] bert-type-vocab-size: 2
[2026-09-26 14:45:10] [config] build-info: ""
[2026-09-26 14:45:10] [config] c

## Let's Check Output Configuration Files and Models

In [123]:
!tree ./model.tf.syl

./model.tf.syl
├── en-my.config.yml
├── model.iter10000.npz
├── model.iter15000.npz
├── model.iter20000.npz
├── model.iter25000.npz
├── model.iter30000.npz
├── model.iter35000.npz
├── model.iter5000.npz
├── model.npz
├── model.npz.decoder.yml
├── model.npz.optimizer.npz
├── model.npz.progress.yml
├── model.npz.yml
├── tf.en-my.log
├── train.log
├── valid.en-my.output
└── valid.log

1 directory, 17 files


In [124]:
!cat ./model.tf.syl/en-my.config.yml

# Marian configuration file generated at 2026-09-26 14:28:29 +0700 with version v1.12.0 65bf82ff 2023-02-21 09:56:29 -0800
# General options
authors: false
cite: false
build-info: ""
workspace: 1000
log: model.tf.syl/train.log
log-level: info
log-time-zone: ""
quiet: false
quiet-translation: true
seed: 1111
check-nan: false
interpolate-env-vars: false
relative-paths: false
sigterm: save-and-exit
# Model options
model: model.tf.syl/model.npz
pretrained-model: ""
ignore-model-config: false
type: transformer
dim-vocabs:
  - 0
  - 0
dim-emb: 512
factors-dim-emb: 0
factors-combine: sum
lemma-dependency: ""
lemma-dim-emb: 0
dim-rnn: 1024
enc-type: bidirectional
enc-cell: gru
enc-cell-depth: 1
enc-depth: 2
dec-cell: gru
dec-cell-base-depth: 2
dec-cell-high-depth: 1
dec-depth: 2
skip: false
layer-normalization: false
right-left: false
input-types:
  []
best-deep: false
tied-embeddings: true
tied-embeddings-src: false
tied-embeddings-all: false
output-omit-bias: false
transformer-heads: 8
trans

In [125]:
!tree ./model.tf.es10.syl/

./model.tf.es10.syl/
├── en-my.config.yml
├── model.iter10000.npz
├── model.iter15000.npz
├── model.iter20000.npz
├── model.iter25000.npz
├── model.iter30000.npz
├── model.iter35000.npz
├── model.iter40000.npz
├── model.iter45000.npz
├── model.iter50000.npz
├── model.iter5000.npz
├── model.iter55000.npz
├── model.iter60000.npz
├── model.npz
├── model.npz.decoder.yml
├── model.npz.optimizer.npz
├── model.npz.progress.yml
├── model.npz.yml
├── tf.en-my.log
├── train.log
├── valid.en-my.output
└── valid.log

1 directory, 22 files


## Let's Play Hyperparameters

    --valid-mini-batch 128 

In [ ]:
!./tf.syl.play.sh

[2026-09-26 15:34:22] [marian] Marian v1.12.0 65bf82ff 2023-02-21 09:56:29 -0800
[2026-09-26 15:34:22] [marian] Running on lst-hpc3090 as process 1574181 with command line:
[2026-09-26 15:34:22] [marian] marian -c model.tf.play.syl/en-my.config.yml
[2026-09-26 15:34:22] [config] after: 0e
[2026-09-26 15:34:22] [config] after-batches: 0
[2026-09-26 15:34:22] [config] after-epochs: 0
[2026-09-26 15:34:22] [config] all-caps-every: 0
[2026-09-26 15:34:22] [config] allow-unk: false
[2026-09-26 15:34:22] [config] authors: false
[2026-09-26 15:34:22] [config] beam-size: 6
[2026-09-26 15:34:22] [config] bert-class-symbol: "[CLS]"
[2026-09-26 15:34:22] [config] bert-mask-symbol: "[MASK]"
[2026-09-26 15:34:22] [config] bert-masking-fraction: 0.15
[2026-09-26 15:34:22] [config] bert-sep-symbol: "[SEP]"
[2026-09-26 15:34:22] [config] bert-train-type-embeddings: true
[2026-09-26 15:34:22] [config] bert-type-vocab-size: 2
[2026-09-26 15:34:22] [config] build-info: ""
[2026-09-26 15:34:22] [config] c

## Play Hyperparameter 2

model_folder="model.tf.play2.syl";  
    --valid-mini-batch 128 \
   --enc-depth 3 --dec-depth 3 \  
    --transformer-heads 10 \

In [129]:
!./tf.syl.play.sh

[2026-09-26 16:10:28] [marian] Marian v1.12.0 65bf82ff 2023-02-21 09:56:29 -0800
[2026-09-26 16:10:28] [marian] Running on lst-hpc3090 as process 1586239 with command line:
[2026-09-26 16:10:28] [marian] marian -c model.tf.play2.syl/en-my.config.yml
[2026-09-26 16:10:28] [config] after: 0e
[2026-09-26 16:10:28] [config] after-batches: 0
[2026-09-26 16:10:28] [config] after-epochs: 0
[2026-09-26 16:10:28] [config] all-caps-every: 0
[2026-09-26 16:10:28] [config] allow-unk: false
[2026-09-26 16:10:28] [config] authors: false
[2026-09-26 16:10:28] [config] beam-size: 6
[2026-09-26 16:10:28] [config] bert-class-symbol: "[CLS]"
[2026-09-26 16:10:28] [config] bert-mask-symbol: "[MASK]"
[2026-09-26 16:10:28] [config] bert-masking-fraction: 0.15
[2026-09-26 16:10:28] [config] bert-sep-symbol: "[SEP]"
[2026-09-26 16:10:28] [config] bert-train-type-embeddings: true
[2026-09-26 16:10:28] [config] bert-type-vocab-size: 2
[2026-09-26 16:10:28] [config] build-info: ""
[2026-09-26 16:10:28] [config] 

**The above is just error example!**

## Testing with Built Models

In [130]:
!marian-decoder --help

Marian: Fast Neural Machine Translation in C++
Usage: marian-decoder [OPTIONS]

General options:
  -h,--help                             Print this help message and exit
  --version                             Print the version number and exit
  --authors                             Print list of authors and exit
  --cite                                Print citation and exit
  --build-info TEXT                     Print CMake build options and exit. Set to 'all' to print advanced options
  -c,--config VECTOR ...                Configuration file(s). If multiple, later overrides earlier
  -w,--workspace INT=512                Preallocate arg MB of work space. Negative `--workspace -N` value allocates workspace as total available GPU memory minus N megabytes.
  --log TEXT                            Log training process information to file given by arg
  --log-level TEXT=info                 Set verbosity level of logging: trace, debug, info, warn, err(or), critical, off
  --log-time-zon

In [131]:
%cd /mnt/disk1/ye/exp/alt-tran/model.tf.syl

/mnt/disk1/ye/exp/alt-tran/model.tf.syl


In [132]:
!cat ./test-eval.sh

#!/bin/bash

for i in {5000..35000..5000}
do
    marian-decoder -m ./model.iter$i.npz -v /mnt/disk1/ye/exp/alt-tran/data/vocab/en_vocab.yml /mnt/disk1/ye/exp/alt-tran/data/vocab/my_syl_vocab.yml --devices 0 --output hyp.iter$i.my < /mnt/disk1/ye/exp/alt-tran/data/syl/test.en;
    echo "Evaluation with hyp.iter$i.my, Transformer model:" >> eval-result.txt;
    perl /home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/generic/multi-bleu.perl /mnt/disk1/ye/exp/alt-tran/data/syl/test.my < ./hyp.iter$i.my >> eval-result.txt;

done


In [133]:
!time ./test-eval.sh

[2026-09-26 16:24:25] [marian] Marian v1.12.0 65bf82ff 2023-02-21 09:56:29 -0800
[2026-09-26 16:24:25] [marian] Running on lst-hpc3090 as process 1590753 with command line:
[2026-09-26 16:24:25] [marian] marian-decoder -m ./model.iter5000.npz -v /mnt/disk1/ye/exp/alt-tran/data/vocab/en_vocab.yml /mnt/disk1/ye/exp/alt-tran/data/vocab/my_syl_vocab.yml --devices 0 --output hyp.iter5000.my
[2026-09-26 16:24:25] [config] alignment: ""
[2026-09-26 16:24:25] [config] allow-special: false
[2026-09-26 16:24:25] [config] allow-unk: false
[2026-09-26 16:24:25] [config] authors: false
[2026-09-26 16:24:25] [config] beam-size: 12
[2026-09-26 16:24:25] [config] bert-class-symbol: "[CLS]"
[2026-09-26 16:24:25] [config] bert-mask-symbol: "[MASK]"
[2026-09-26 16:24:25] [config] bert-masking-fraction: 0.15
[2026-09-26 16:24:25] [config] bert-sep-symbol: "[SEP]"
[2026-09-26 16:24:25] [config] bert-train-type-embeddings: true
[2026-09-26 16:24:25] [config] bert-type-vocab-size: 2
[2026-09-26 16:24:25] [co

IOPub data rate exceeded.
The Jupyter server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--ServerApp.iopub_data_rate_limit`.

Current values:
ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
ServerApp.rate_limit_window=3.0 (secs)



[2026-09-26 16:32:52] Best translation 1001 : လေ့ လာ မှု အ ပေါ် အ ခြေ ခံ ထား သော နေ ရာ များ ၏ ရာ ပေါင်း များ စွာ ၊ အ စိုး ရ တာ ဝန် ရှိ သူ များ ၊ စက် ရုံ လုပ် သား များ ၊ စက် ရုံ နှင့် အ လုပ် သ မား များ ၊ အ လုပ် သ မား များ နှင့် ပတ် သက် ၍ ကုမ္ပ ဏီ များ အား လုံး ၏ ကိုယ် စား လှယ် များ သည် ပို မို အ လုပ် သ မား များ ကို သူ တို့ ၏ အ ခြေ အ နေ တွင် ထား ရန် အ ကြံ ပြု သည် ။
[2026-09-26 16:32:52] Best translation 1002 : အ ဓိ က လက် လီ အ ရောင်း ဆိုင် များ တွင် အ ခြေ စိုက် စ ခန်း ၅ ၀ , ၀ ၀ ၀ ကျော် ၏ ထောက် ပံ့ မှု တစ် ခု ရှိ နိုင် ပြီး ၊ ၄င်း သည် ပို မို ထိ ရောက် သော ကုမ္ပ ဏီ များ နှင့် ဆက် သွယ် မှု ပို မို ခက် ခဲ စေ ရန် ခက် ခဲ သည် ။
[2026-09-26 16:32:52] Best translation 1003 : သန်း ခေါင် စာ ရင်း ရ လဒ် များ သည် ယ နေ့ အ ကျယ် တ ဝင့် လူ ဦး ရေ ၏ လူ ဦး ရေ ကို ဖော် ထုတ် သော ၊ အိန္ဒိ ယ နိုင် ငံ ၏ အ ကြီး စား လူ ဦး ရေ တွင် ရပ် တည် လျက် ရှိ သည် ။
[2026-09-26 16:32:52] Best translation 1004 : ဝါ ရှင် တန် သ တင်း စာ အ ဆို အ ရ ၊ အိန္ဒိ ယ နိုင် ငံ ၏ မိ ဘ များ စု မှာ ကိုယ် ဝန် ဆောင် ခဲ့ ပြီး ငယ် ရွယ် သော အ ကြောင်း ပ

In [134]:
!ls

en-my.config.yml  model.iter10000.npz	 model.npz.optimizer.npz
eval-result.txt   model.iter15000.npz	 model.npz.progress.yml
hyp.iter10000.my  model.iter20000.npz	 model.npz.yml
hyp.iter15000.my  model.iter25000.npz	 test-eval.sh
hyp.iter20000.my  model.iter30000.npz	 tf.en-my.log
hyp.iter25000.my  model.iter35000.npz	 train.log
hyp.iter30000.my  model.iter5000.npz	 valid.en-my.output
hyp.iter35000.my  model.npz		 valid.log
hyp.iter5000.my   model.npz.decoder.yml


In [135]:
!cat eval-result.txt

Evaluation with hyp.iter5000.my, Transformer model:
BLEU = 12.29, 52.2/26.6/14.5/7.6 (BP=0.620, ratio=0.677, hyp_len=79892, ref_len=118037)
Evaluation with hyp.iter10000.my, Transformer model:
BLEU = 17.24, 59.3/32.8/18.9/10.8 (BP=0.686, ratio=0.726, hyp_len=85750, ref_len=118037)
Evaluation with hyp.iter15000.my, Transformer model:
BLEU = 19.59, 61.2/34.5/20.2/11.8 (BP=0.736, ratio=0.766, hyp_len=90390, ref_len=118037)
Evaluation with hyp.iter20000.my, Transformer model:
BLEU = 20.35, 61.4/34.7/20.5/12.1 (BP=0.755, ratio=0.780, hyp_len=92099, ref_len=118037)
Evaluation with hyp.iter25000.my, Transformer model:
BLEU = 20.84, 61.7/35.1/20.8/12.3 (BP=0.764, ratio=0.788, hyp_len=92977, ref_len=118037)
Evaluation with hyp.iter30000.my, Transformer model:
BLEU = 21.28, 61.4/35.0/20.9/12.4 (BP=0.780, ratio=0.801, hyp_len=94503, ref_len=118037)
Evaluation with hyp.iter35000.my, Transformer model:
BLEU = 21.20, 61.5/35.2/21.0/12.5 (BP=0.773, ratio=0.795, hyp_len=93895, ref_len=118037)


## Compare Source, Reference and Hypothesis

In [140]:
%cd /mnt/disk1/ye/exp/alt-tran/data/syl/

/mnt/disk1/ye/exp/alt-tran/data/syl


## Source or Input

In [141]:
!head ./test.en

In her statement to MPs , Ruth Kelly confirmed that no banking information was contained within the lost data , nor were individuals &apos; dates of birth .
Senator Joe Ludwig ( Labor , Queensland ) asked of the Minister for Justice and Customs Senator Chris Ellison in Question Time whether he stood by his claim that he made yesterday that the discovery of the Jian Seng demonstrated that the Australian Government had in place &quot; aerial and maritime surveillance to intercept a vessel in these circumstances , and that was done &quot; , and asked why it took the Government two weeks to intercept the vessel after it entered Australian waters .
The cases are the first infections of race horses , despite infecting dozens of recreational horses across NSW and Queensland .
City officials in seven United States cities were warned today about a possible attack on NFL stadiums using &quot; dirty bombs . &quot;
In response to the threat , several teams are scheduled to sweep some stadiums with

## Reference or Target

In [142]:
!head ./test.my

လွှတ် တော် အ မတ် များ သို့ သူ မ ၏ ကြေ ငြာ ချက် တွင် ရုသ် ကယ် လီ က ပျောက် ဆုံး ခဲ့ သော အ ချက် အ လက် များ အ တွင်း မည် သည့် ဘဏ် လုပ် ငန်း သ တင်း အ ချက် အ လက် မှ မ ပါ ၀င် သ ကဲ့ သို့ တစ် ဦး ချင်း စီ ၏ မွေး သက္က ရာဇ် များ လည်း မ ပါ ၀င် ကြောင်း အ တည် ပြု သွား ခဲ့ သည် ။
အ ထက် လွှတ် တော် အ မတ် ဂျိုး လူး ၀တ် ( အ လုပ် သ မား ရေး ရာ ၊ ကွင်း စ လန် ) က တ ရား ရေး ဝန် ကြီး နှင့် အ ကောက် ခွန် အ ထက် လွှတ် တော် အ မတ် ချ ရစ် အယ် လစ် ဆန် အား သြ စ တြေး လျ အ စိုး ရ ပိုင် နက် တွင် “ ဤ အ ခြေ အ နေ များ တွင် သင်္ဘော တစ် စီး အား ဖမ်း ဆီး ရန်ေ ၀ ဟင် နှင့် ရေ ကြောင်း စောင့် ကြည့် မှု နှင့် ထို အ လုပ် မှာ ပြီး ဆုံး ခဲ့ ကြောင်း ” ကို သ ရုပ် ပြ ခဲ့ သည့် ဂျီ အန်း ဆန် ရှာ ဖွေ မှု နှင့် ပတ် သက် ၍ မ နေ့ က သူ ပြု လုပ် ခဲ့ သည် ဆို သော သူ ၏ တောင်း ဆို ချက် အ တွက် သူ ရပ် တည် ခြင်း ရှိ မ ရှိ ကို မေး ခွန်း မေး ချိန် တွင် မေး ခဲ့ ပြီး အ စိုး ရ က သြ စ တြေး လျ ရေ ကြောင်း သို့ ၀င် ရောက် ခဲ့ ပြီး နောက် သင်္ဘော ကို ဖမ်း ဆီး ရန် ကိစ္စ အား မည် သည့် အ တွက် ကြောင့် အ ချိန် နှစ် ပတ် ယူ ခဲ့ သည် ကို မေး ခဲ့ သည် ။
အန် အက် (စ် ) ဒ ဗ လျူ နှင့် 

## Hypothesis or Model hyp.iter30000.my's Translation

In [143]:
!head /mnt/disk1/ye/exp/alt-tran/model.tf.syl/hyp.iter30000.my


လွတ် တော် အ မတ် ဂျိုး ဆန် ( ကွင်း စ လန်း အ လုပ် သ မား ) က ဆက် လက် ပြု လုပ် ခဲ့ ပြီး တ ရား မျှ တ မှု အ တွက် ဝန် ကြီး အ ထက် လွှတ် တော် အ မတ် ခ ရစ် ဆန် သည် သူ ၏ သင်္ဘော ကို ရှာ ဖွေ တွေ့ ရှိ ရန် တောင်း ဆို ခဲ့ ပြီး နောက် မ နေ့ က “ သြ စ တြေး လျ ရေ ပိုင် နက် ၏ အ ခြေ အ နေ များ တွင် ထို ရှာ ဖွေ တွေ့ ရှိ မှု သည် အ စိုး ရ နှင့် ရေ ပိုင် နက် ထဲ သို့ ဝင် ရောက် ခဲ့ ပြီး ထို ရှာ ဖွေ တွေ့ ရှိ မှု တစ် ခု ပြု လုပ် ခဲ့ သည် ” ဟု ပြော ကြား ခဲ့ သည် ။
ကူး စက် ခံ ရ သော မြင်း များ သည် အန် အက်စ် ဒ ဗ လျူ နှင့် ကွင်း စ လန်း အ ပန်း ဖြေ ရေး ၏ ပ ထ မ ဦး ဆုံး အ ကြောင်း ဖြစ် သော် လည်း ၊ အ ဆို ပါ ပြိုင် ပွဲ သည် အန် အက်စ် ဒ ဗ လျူ နှင့် ကွင်း စ လန်း ပြည် နယ် အ နှံ့ မြင်း များ ဖြစ် ကြ သည် ။
အ မေ ရိ ကန် ပြည် ထောင် စု ရှိ မြို့ ခု နှစ် ခု တွင် မြို့ တော် အ ရာ ရှိ များ မှာ အား က စား ကွင်း လယ် သ မား “ အား က စား ကွင်း ” မှာ ဖြစ် နိုင် သော တိုက် ခိုက် မှု တစ် ခု နှင့် ပတ် သက် ၍ သ တိ ပေး ခဲ့ သည် ။
တုန့် ပြန် မှု တွင် ၊ များ စွာ သော အ ဆို ပါ ခြိမ်း ခြောက် မှု သည် အ သင်း များ ၏ တည် ရှိ မှု အ တွက် တ နင်္ဂ နွေ နေ့ တွင် အား က စား ကွ

## Prepare Shell Script for Early-Stop 10 Models

In [136]:
%cd /mnt/disk1/ye/exp/alt-tran/model.tf.es10.syl

/mnt/disk1/ye/exp/alt-tran/model.tf.es10.syl


In [137]:
!cat test-eval.sh

#!/bin/bash

for i in {5000..60000..5000}
do
    marian-decoder -m ./model.iter$i.npz -v /mnt/disk1/ye/exp/alt-tran/data/vocab/en_vocab.yml /mnt/disk1/ye/exp/alt-tran/data/vocab/my_syl_vocab.yml --devices 0 --output hyp.iter$i.my < /mnt/disk1/ye/exp/alt-tran/data/syl/test.en;
    echo "Evaluation with hyp.iter$i.my, Transformer model:" >> eval-result.txt;
    perl /home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/generic/multi-bleu.perl /mnt/disk1/ye/exp/alt-tran/data/syl/test.my < ./hyp.iter$i.my >> eval-result.txt;

done


**Check GPU Status During Decoding Time**

In [139]:
!nvidia-smi

Sat Sep 26 16:49:28 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 595.84                 Driver Version: 595.84         CUDA Version: 13.2     |
+-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 3090 Ti     Off |   00000000:01:00.0 Off |                  Off |
| 32%   64C    P2            193W /  480W |    4748MiB /  24564MiB |     66%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

In [146]:
%cd /mnt/disk1/ye/exp/alt-tran/model.tf.es10.syl

/mnt/disk1/ye/exp/alt-tran/model.tf.es10.syl


In [147]:
!cat eval-result.txt

Evaluation with hyp.iter5000.my, Transformer model:
BLEU = 12.34, 52.3/26.7/14.5/7.7 (BP=0.620, ratio=0.677, hyp_len=79896, ref_len=118037)
Evaluation with hyp.iter10000.my, Transformer model:
BLEU = 17.42, 59.3/32.8/19.0/10.8 (BP=0.692, ratio=0.731, hyp_len=86289, ref_len=118037)
Evaluation with hyp.iter15000.my, Transformer model:
BLEU = 19.48, 60.8/34.2/20.1/11.7 (BP=0.737, ratio=0.766, hyp_len=90431, ref_len=118037)
Evaluation with hyp.iter20000.my, Transformer model:
BLEU = 20.44, 61.4/34.8/20.6/12.1 (BP=0.757, ratio=0.782, hyp_len=92306, ref_len=118037)
Evaluation with hyp.iter25000.my, Transformer model:
BLEU = 21.03, 61.5/35.0/20.8/12.3 (BP=0.772, ratio=0.795, hyp_len=93818, ref_len=118037)
Evaluation with hyp.iter30000.my, Transformer model:
BLEU = 21.18, 61.3/35.0/20.8/12.4 (BP=0.777, ratio=0.798, hyp_len=94248, ref_len=118037)
Evaluation with hyp.iter35000.my, Transformer model:
BLEU = 21.09, 61.3/35.0/20.8/12.3 (BP=0.774, ratio=0.796, hyp_len=93948, ref_len=118037)
Evaluati

## Preparing Shell Script for Transformer Model (with Myanmar Word Unit)

In [150]:
%cd /mnt/disk1/ye/exp/alt-tran/

/mnt/disk1/ye/exp/alt-tran


```bash
#!/bin/bash

## Written by Ye Kyaw Thu, Language Understanding Lab, AI Mario, Myanmar
## Experiments for AI Engineering Fundamental Class (Batch-2)
## Date: 27 Sept 2026

## Old Notes
#     --mini-batch-fit -w 10000 --maxi-batch 1000 \
#    --mini-batch-fit -w 1000 --maxi-batch 100 \
#     --tied-embeddings-all \
#     --tied-embeddings \
#     --valid-metrics cross-entropy perplexity translation bleu \
#     --transformer-dropout 0.1 --label-smoothing 0.1 \
#     --learn-rate 0.0003 --lr-warmup 16000 --lr-decay-inv-sqrt 16000 --lr-report \
#     --optimizer-params 0.9 0.98 1e-09 --clip-norm 5 \

model_folder="model.tf.word";
mkdir -p ${model_folder};
data_path="/mnt/disk1/ye/exp/alt-tran/data/word/";
vocab_path="/mnt/disk1/ye/exp/alt-tran/data/vocab/"
src="en"; tgt="my";

marian \
    --model ${model_folder}/model.npz --type transformer \
    --train-sets ${data_path}/train.${src} ${data_path}/train.${tgt} \
    --max-length 200 \
    --vocabs  ${vocab_path}/en_vocab.yml  ${vocab_path}/my_vocab.yml \
    --mini-batch-fit -w 1000 --maxi-batch 100 \
    --early-stopping 5 \
    --valid-freq 5000 --save-freq 5000 --disp-freq 500 \
    --valid-metrics cross-entropy perplexity bleu \
    --valid-sets ${data_path}/valid.${src} ${data_path}/valid.${tgt} \
    --valid-translation-output ${model_folder}/valid.${src}-${tgt}.output \
    --quiet-translation \
    --valid-mini-batch 64 \
    --beam-size 6 --normalize 0.6 \
    --log ${model_folder}/train.log \
    --valid-log ${model_folder}/valid.log \
    --enc-depth 2 --dec-depth 2 \
    --transformer-heads 8 \
    --transformer-postprocess-emb d \
    --transformer-postprocess dan \
    --transformer-dropout 0.3 --label-smoothing 0.1 \
    --learn-rate 0.0003 --lr-warmup 0 --lr-decay-inv-sqrt 16000 --lr-report \
    --clip-norm 5 \
    --tied-embeddings \
    --devices 0 --sync-sgd --seed 1111 \
    --exponential-smoothing \
    --dump-config > ${model_folder}/${src}-${tgt}.config.yml

time marian -c ${model_folder}/${src}-${tgt}.config.yml 2>&1 | tee ${model_folder}/tf.${src}-${tgt}.log

```

## Training English-Myanmar Transformer Model with Word Unit

In [151]:
!./tf.word.sh

[2026-09-27 21:16:14] [marian] Marian v1.12.0 65bf82ff 2023-02-21 09:56:29 -0800
[2026-09-27 21:16:14] [marian] Running on lst-hpc3090 as process 1976217 with command line:
[2026-09-27 21:16:14] [marian] marian -c model.tf.word/en-my.config.yml
[2026-09-27 21:16:14] [config] after: 0e
[2026-09-27 21:16:14] [config] after-batches: 0
[2026-09-27 21:16:14] [config] after-epochs: 0
[2026-09-27 21:16:14] [config] all-caps-every: 0
[2026-09-27 21:16:14] [config] allow-unk: false
[2026-09-27 21:16:14] [config] authors: false
[2026-09-27 21:16:14] [config] beam-size: 6
[2026-09-27 21:16:14] [config] bert-class-symbol: "[CLS]"
[2026-09-27 21:16:14] [config] bert-mask-symbol: "[MASK]"
[2026-09-27 21:16:14] [config] bert-masking-fraction: 0.15
[2026-09-27 21:16:14] [config] bert-sep-symbol: "[SEP]"
[2026-09-27 21:16:14] [config] bert-train-type-embeddings: true
[2026-09-27 21:16:14] [config] bert-type-vocab-size: 2
[2026-09-27 21:16:14] [config] build-info: ""
[2026-09-27 21:16:14] [config] check

## Check Output Models

In [152]:
!ls ./model.tf.word/ -lh

total 1.8G
-rw-rw-r-- 1 ye ye 4.8K Sep 27 21:16 en-my.config.yml
-rw-rw-r-- 1 ye ye 162M Sep 27 21:38 model.iter10000.npz
-rw-rw-r-- 1 ye ye 162M Sep 27 21:50 model.iter15000.npz
-rw-rw-r-- 1 ye ye 162M Sep 27 22:02 model.iter20000.npz
-rw-rw-r-- 1 ye ye 162M Sep 27 22:13 model.iter25000.npz
-rw-rw-r-- 1 ye ye 162M Sep 27 22:25 model.iter30000.npz
-rw-rw-r-- 1 ye ye 162M Sep 27 21:26 model.iter5000.npz
-rw-rw-r-- 1 ye ye 162M Sep 27 22:26 model.npz
-rw-rw-r-- 1 ye ye  273 Sep 27 22:26 model.npz.decoder.yml
-rw-rw-r-- 1 ye ye 646M Sep 27 22:26 model.npz.optimizer.npz
-rw-rw-r-- 1 ye ye  14K Sep 27 22:26 model.npz.progress.yml
-rw-rw-r-- 1 ye ye 4.6K Sep 27 22:26 model.npz.yml
-rwxrwxr-x 1 ye ye  529 Sep 27 22:28 test-eval.sh
-rw-rw-r-- 1 ye ye  60K Sep 27 22:26 tf.en-my.log
-rw-rw-r-- 1 ye ye  60K Sep 27 22:26 train.log
-rw-rw-r-- 1 ye ye 1.8M Sep 27 22:26 valid.en-my.output
-rw-rw-r-- 1 ye ye 4.4K Sep 27 22:26 valid.log


## Prepare a Shell Script for Testing

In [153]:
!cat ./model.tf.word/test-eval.sh

#!/bin/bash

for i in {5000..30000..5000}
do
    marian-decoder -m ./model.iter$i.npz -v /mnt/disk1/ye/exp/alt-tran/data/vocab/en_vocab.yml /mnt/disk1/ye/exp/alt-tran/data/vocab/my_vocab.yml --devices 0 --output hyp.iter$i.my < /mnt/disk1/ye/exp/alt-tran/data/word/test.en;
    echo "Evaluation with hyp.iter$i.my, Transformer Word model:" >> eval-result.txt;
    perl /home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/generic/multi-bleu.perl /mnt/disk1/ye/exp/alt-tran/data/word/test.my < ./hyp.iter$i.my >> eval-result.txt;

done


## Testing/Evaluation with Word Model

In [155]:
%cd ./model.tf.word/

/mnt/disk1/ye/exp/alt-tran/model.tf.word


In [ ]:
!time ./test-eval.sh

[2026-09-27 22:37:06] [marian] Marian v1.12.0 65bf82ff 2023-02-21 09:56:29 -0800
[2026-09-27 22:37:06] [marian] Running on lst-hpc3090 as process 1993653 with command line:
[2026-09-27 22:37:06] [marian] marian-decoder -m ./model.iter5000.npz -v /mnt/disk1/ye/exp/alt-tran/data/vocab/en_vocab.yml /mnt/disk1/ye/exp/alt-tran/data/vocab/my_vocab.yml --devices 0 --output hyp.iter5000.my
[2026-09-27 22:37:06] [config] alignment: ""
[2026-09-27 22:37:06] [config] allow-special: false
[2026-09-27 22:37:06] [config] allow-unk: false
[2026-09-27 22:37:06] [config] authors: false
[2026-09-27 22:37:06] [config] beam-size: 12
[2026-09-27 22:37:06] [config] bert-class-symbol: "[CLS]"
[2026-09-27 22:37:06] [config] bert-mask-symbol: "[MASK]"
[2026-09-27 22:37:06] [config] bert-masking-fraction: 0.15
[2026-09-27 22:37:06] [config] bert-sep-symbol: "[SEP]"
[2026-09-27 22:37:06] [config] bert-train-type-embeddings: true
[2026-09-27 22:37:06] [config] bert-type-vocab-size: 2
[2026-09-27 22:37:06] [config

In [157]:
!cat eval-result.txt

Evaluation with hyp.iter5000.my, Transformer Word model:
BLEU = 6.41, 44.4/16.8/6.5/2.3 (BP=0.623, ratio=0.678, hyp_len=53561, ref_len=78949)
Evaluation with hyp.iter10000.my, Transformer Word model:
BLEU = 8.64, 47.7/18.9/7.7/3.0 (BP=0.718, ratio=0.751, hyp_len=59304, ref_len=78949)
Evaluation with hyp.iter15000.my, Transformer Word model:
BLEU = 9.38, 48.3/19.3/8.1/3.3 (BP=0.748, ratio=0.775, hyp_len=61181, ref_len=78949)
Evaluation with hyp.iter20000.my, Transformer Word model:
BLEU = 9.82, 48.5/19.5/8.3/3.5 (BP=0.761, ratio=0.785, hyp_len=61991, ref_len=78949)
Evaluation with hyp.iter25000.my, Transformer Word model:
BLEU = 9.92, 48.3/19.5/8.2/3.4 (BP=0.778, ratio=0.800, hyp_len=63126, ref_len=78949)
Evaluation with hyp.iter30000.my, Transformer Word model:
BLEU = 10.27, 48.7/20.1/8.7/3.7 (BP=0.774, ratio=0.796, hyp_len=62821, ref_len=78949)


## Syllable Breaking on Word Hyp Files


အထက်က syllable unit နဲ့ ဆောက်ထားတဲ့ မော်ဒယ်နဲ့ နှိုင်းယှဉ်ဖို့အတွက်က လက်ရှိ word model ရဲ့ ရလဒ်တွေကို syllable breaking လုပ်ပေးဖို့ လိုအပ်ပါတယ်။ အဲဒါမှ နှိုင်းယှဉ်လို့ ရမှာမို့ပါ။  

In [158]:
%pwd

'/mnt/disk1/ye/exp/alt-tran/model.tf.word'

In [159]:
!mkdir syl

In [161]:
!cp hyp*.my ./syl/

In [162]:
!ls ./syl/*

./syl/hyp.iter10000.my	./syl/hyp.iter20000.my	./syl/hyp.iter30000.my
./syl/hyp.iter15000.my	./syl/hyp.iter25000.my	./syl/hyp.iter5000.my


In [163]:
!head -n 3 ./syl/*

==> ./syl/hyp.iter10000.my <==
သူ မ၏ လွှတ်တော် ထဲ တွင် ၊ ကယ် လီ ဖိုး နီး ယား သည် မွေးဖွား ခဲ့ သော သတင်း အချက်အလက် များ ပါဝင် ခြင်း မရှိ ခြင်း ဟု အတည်ပြု ခဲ့ ပြီး ၊ သူ မ၏ မွေးဖွား သော သတင်း အချက်အလက် များ ကြောင့် သေဆုံး ခဲ့ သည် ။
တက် စီ အန် ဘီစီ ၊ ဂျိုး လူး ဝစ် က ၀န် ကြီး ချုပ် ( သြစတေးလျ အလုပ်သမား ရေးရာ ဝန်ကြီး ၊ ခရစ် စ မန်း ) က ၊ သူ သည် “ ဩ စ တြေး လျ အ စိုး ရ၏ လုပ်ငန်းစဉ် များ ကို ရှာဖွေ တွေ့ ရှိ ခဲ့ ပြီးနောက် မနေ့ က သြစတြေးလျ ရေ ပိုင်နက် ထဲ မှာ မှားယွင်း သော အခြေအနေ တစ် ခု ဖြစ် ခဲ့ သည် ဟု အခိုင်အမာ ပြောဆို ခဲ့ သည် ။ ”
ကူးစက် ခံ ရ သော မြင်း များ သည် မြင်း များ ဖြစ် သော်လည်း ၊ ကွင်းစ လန်း နှင့် ကွင်းစ လန်း စတိုင် နာ တို့ ဖြစ် ကြ သည် ။

==> ./syl/hyp.iter15000.my <==
သူ မ၏ လွှတ်တော် အမတ် အမ် ပီ ဘ ရိုင် ယန် သည် မွေးဖွား ခဲ့ သော သတင်း အချက်အလက် များ အတွင်း သတင်း အချက်အလက် များ ပါဝင် ခြင်း မရှိ ဟု အတည်ပြု ခဲ့ သည် ။
ဩ စ တြေး လျ အလုပ်သမား ရေးရာ ဝန်ကြီး ဂျိုး လူး ဝစ် က ၊ သူ သည် အကောက်ခွန် နှင့် အကောက်ခွန် အဖွဲ့ အ စည်း၏ ၀န် ကြီး ချုပ် ခရစ် စ တြေး လျ အစိုးရ ဝန်ကြီး အန် ဒ ရော့ ဆန် က ၊ သူ သည် “ ၎

In [165]:
%cd syl

/mnt/disk1/ye/exp/alt-tran/model.tf.word/syl


In [166]:
!cat break_all.sh

#!/bin/bash

for f in hyp.iter*.my; do
    python /mnt/disk1/ye/exp/alt-tran/sylbreak4all.py -i "$f" -s " " > "${f}.syl"
done


In [167]:
!./break_all.sh

## Manual Checking on Syllable Segmented Outputs

In [168]:
!head -n 3 *.syl

==> hyp.iter10000.my.syl <==
သူ မ ၏ လွှတ် တော် ထဲ တွင် ၊ ကယ် လီ ဖိုး နီး ယား သည် မွေး ဖွား ခဲ့ သော သ တင်း အ ချက် အ လက် များ ပါ ဝင် ခြင်း မ ရှိ ခြင်း ဟု အ တည် ပြု ခဲ့ ပြီး ၊ သူ မ ၏ မွေး ဖွား သော သ တင်း အ ချက် အ လက် များ ကြောင့် သေ ဆုံး ခဲ့ သည် ။
တက် စီ အန် ဘီ စီ ၊ ဂျိုး လူး ဝစ် က ၀န် ကြီး ချုပ် ( သြ စ တေး လျ အ လုပ် သ မား ရေး ရာ ဝန် ကြီး ၊ ခ ရစ် စ မန်း ) က ၊ သူ သည် “ ဩ စ တြေး လျ အ စိုး ရ ၏ လုပ် ငန်း စဉ် များ ကို ရှာ ဖွေ တွေ့ ရှိ ခဲ့ ပြီး နောက် မ နေ့ က သြ စ တြေး လျ ရေ ပိုင် နက် ထဲ မှာ မှား ယွင်း သော အ ခြေ အ နေ တစ် ခု ဖြစ် ခဲ့ သည် ဟု အ ခိုင် အ မာ ပြော ဆို ခဲ့ သည် ။ ”
ကူး စက် ခံ ရ သော မြင်း များ သည် မြင်း များ ဖြစ် သော် လည်း ၊ ကွင်း စ လန်း နှင့် ကွင်း စ လန်း စ တိုင် နာ တို့ ဖြစ် ကြ သည် ။

==> hyp.iter15000.my.syl <==
သူ မ ၏ လွှတ် တော် အ မတ် အမ် ပီ ဘ ရိုင် ယန် သည် မွေး ဖွား ခဲ့ သော သ တင်း အ ချက် အ လက် များ အ တွင်း သ တင်း အ ချက် အ လက် များ ပါ ဝင် ခြင်း မ ရှိ ဟု အ တည် ပြု ခဲ့ သည် ။
ဩ စ တြေး လျ အ လုပ် သ မား ရေး ရာ ဝန် ကြီး ဂျိုး လူး ဝစ် က ၊ သူ သည် အ ကောက် ခွန် နှင့် အ ကောက် ခွန် အ ဖွဲ့ အ စည်း ၏

## Prepare a New Evaluation Shell Script

In [169]:
!cat ./eval_syl.sh

#!/bin/bash

for i in {5000..30000..5000}
do
    echo "Evaluation with hyp.iter$i.my.syl, Transformer Word model (Syllable):" >> eval-result.txt;
    perl /home/ye/tool/mosesbin/ubuntu-17.04/moses/scripts/generic/multi-bleu.perl /mnt/disk1/ye/exp/alt-tran/data/syl/test.my < ./hyp.iter$i.my.syl >> eval-result.txt;

done



In [170]:
%pwd

'/mnt/disk1/ye/exp/alt-tran/model.tf.word/syl'

In [171]:
!./eval_syl.sh

In [172]:
!ls -lh ./eval-result.txt

-rw-rw-r-- 1 ye ye 966 Sep 28 00:37 ./eval-result.txt


In [173]:
!cat ./eval-result.txt

Evaluation with hyp.iter5000.my.syl, Transformer Word model (Syllable):
BLEU = 11.45, 51.7/25.5/13.7/7.3 (BP=0.601, ratio=0.663, hyp_len=78239, ref_len=118037)
Evaluation with hyp.iter10000.my.syl, Transformer Word model (Syllable):
BLEU = 15.41, 56.1/29.0/16.1/8.9 (BP=0.700, ratio=0.737, hyp_len=87042, ref_len=118037)
Evaluation with hyp.iter15000.my.syl, Transformer Word model (Syllable):
BLEU = 16.47, 56.9/29.7/16.6/9.3 (BP=0.728, ratio=0.759, hyp_len=89589, ref_len=118037)
Evaluation with hyp.iter20000.my.syl, Transformer Word model (Syllable):
BLEU = 17.03, 57.2/30.0/17.0/9.7 (BP=0.739, ratio=0.767, hyp_len=90583, ref_len=118037)
Evaluation with hyp.iter25000.my.syl, Transformer Word model (Syllable):
BLEU = 17.42, 57.1/30.0/17.0/9.7 (BP=0.756, ratio=0.781, hyp_len=92207, ref_len=118037)
Evaluation with hyp.iter30000.my.syl, Transformer Word model (Syllable):
BLEU = 17.79, 57.4/30.6/17.5/10.1 (BP=0.754, ratio=0.780, hyp_len=92015, ref_len=118037)


## Compare Transformer (Syllable) vs Transformer Word (Syllable) Model Result

| Iteration | Transformer (Syl) BLEU | Transformer Word (Syl) BLEU | Difference / Transformer (Syl) − Transformer Word (Syl) |
|---:|---:|---:|---:|
| 5000 | 12.29 | 11.45 | +0.84 |
| 10000 | 17.24 | 15.41 | +1.83 |
| 15000 | 19.59 | 16.47 | +3.12 |
| 20000 | 20.35 | 17.03 | +3.32 |
| 25000 | 20.84 | 17.42 | +3.42 |
| 30000 | 21.28 | 17.79 | +3.49 |
| 35000 | 21.20 | — | — |
| **Best** | **21.28 @ 30000** | **17.79 @ 30000** | **+3.49 best-vs-best** |

## Best checkpoint comparison

| Model | Best iteration | BLEU | 1-gram | 2-gram | 3-gram | 4-gram | BP | ratio | hyp_len | ref_len |
|---|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| Transformer (Syl) | 30000 | **21.28** | 61.4 | 35.0 | 20.9 | 12.4 | 0.780 | 0.801 | 94503 | 118037 |
| Transformer Word (Syl) | 30000 | **17.79** | 57.4 | 30.6 | 17.5 | 10.1 | 0.754 | 0.780 | 92015 | 118037 |

